{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Fix inconsistencies in parent column\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "from glob import glob\n",
    "import yaml\n",
    "import re\n",
    "import copy\n",
    "import json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hard Coded stuff\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {},
   "outputs": [],
   "source": [
    "validation_coder = {\n",
    "    \"number\": \"regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)\",\n",
    "    \"integer\": \"regex search ([0-9]+)\",\n",
    "    \"string\": \"\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_cols = [\n",
    "    \"Attribute\",\n",
    "    \"Description\",\n",
    "    \"Valid Values\",\n",
    "    \"DependsOn\",\n",
    "    \"Properties\",\n",
    "    \"Required\",\n",
    "    \"Parent\",\n",
    "    \"DependsOn Component\",\n",
    "    \"Source\",\n",
    "    \"Validation Rules\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean list columns into single string\n",
    "def join_strings(string):\n",
    "    try:\n",
    "        return \",\".join(string)\n",
    "    except:\n",
    "        return \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_df(df, pattern):\n",
    "    mask = np.column_stack(\n",
    "        [df[col].str.contains(pattern, na=False, flags=re.IGNORECASE) for col in df]\n",
    "    )\n",
    "\n",
    "    df = df.loc[mask.any(axis=1)]\n",
    "\n",
    "    with pd.option_context(\"display.max_colwidth\", None):\n",
    "        display(df)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_row(df, attribute):\n",
    "    \"\"\"Get indexes of the dataframe\"\"\"\n",
    "    indexes = df.index[\n",
    "        df[\"Attribute\"].str.contains(\n",
    "            \"(^\" + re.escape(attribute) + \"$)\", flags=re.IGNORECASE\n",
    "        )\n",
    "    ].tolist()\n",
    "    if len(indexes) != 0:\n",
    "        return indexes\n",
    "    else:\n",
    "        print(attribute)\n",
    "        return None\n",
    "\n",
    "\n",
    "def replace_valid_value(df, indexes, regex_dict, attribute):\n",
    "    \"\"\"Alter the dataframe valid values with the replacement value\"\"\"\n",
    "    if indexes == None:\n",
    "        return df\n",
    "\n",
    "    elif len(indexes) > 0:\n",
    "        regex_dict = regex_dict[attribute]\n",
    "\n",
    "        for index in indexes:\n",
    "            df.loc[index, \"Valid Values\"] = re.sub(\n",
    "                **regex_dict, string=df.loc[index, \"Valid Values\"]\n",
    "            )\n",
    "\n",
    "            # print(df.loc[index, 'Valid Values'])\n",
    "        return df\n",
    "    else:\n",
    "        return df\n",
    "\n",
    "\n",
    "def code_equals_values(df, regex_dict, attribute):\n",
    "    print(\"attribute: \", attribute)\n",
    "\n",
    "    indexes = find_row(df, attribute)\n",
    "\n",
    "    print(\"Index: \", indexes)\n",
    "\n",
    "    # regex = replace_value(df, replacements, indexes)\n",
    "\n",
    "    # print(\"Regex: \", regex)\n",
    "    if indexes == None:\n",
    "        return df\n",
    "    else:\n",
    "        df = replace_valid_value(df, indexes, regex_dict, attribute)\n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unzip compressed folder if downloaded from Google Drive\n",
    "# %unzip 'RFC Tables-20230620T181152Z-001.zip'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collect RFCs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"local_configs/notebook_config.yaml\", \"r\") as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../_data/RFC Tables/EL Assay_ scRNAseq data model.xlsx',\n",
       " '../_data/RFC Tables/EL RFC_ Individual_ Human data model.xlsx',\n",
       " '../_data/RFC Tables/EL RFC_ Assay_proteomics Data Model.xlsx',\n",
       " '../_data/RFC Tables/EL RFC Assay_Whole Genome Sequencing Data Model.xlsx',\n",
       " '../_data/RFC Tables/EL RFC Microbiome Data Model.xlsx',\n",
       " '../_data/RFC Tables/EL RFC_ Biospecimen_nonHuman data model.xlsx',\n",
       " '../_data/RFC Tables/EL RFC_ Individual_ nonHuman data model.xlsx',\n",
       " '../_data/RFC Tables/EL RFC Metabolomics Human Data Model.xlsx',\n",
       " '../_data/RFC Tables/EL RFC genotyping_assay.xlsx',\n",
       " '../_data/RFC Tables/EL RFC_ Biospecimen_human data model.xlsx',\n",
       " '../_data/RFC Tables/EL Assay_ bsSeq (bisulfite-seq_WGBS_methylseq_methylomics) data model.xlsx',\n",
       " '../_data/RFC Tables/EL Assay_ RNAseq data model.xlsx']"
      ]
     },
     "execution_count": 273,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get all the RFC file paths\n",
    "file_paths = glob(\"../_data/RFC Tables/*\")\n",
    "file_paths"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create Data Model for Schematic\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm = pd.DataFrame()\n",
    "\n",
    "# parse through files to create complete data model\n",
    "for fp in file_paths:\n",
    "    file_name = os.path.basename(fp)\n",
    "\n",
    "    temp = pd.read_excel(fp)\n",
    "\n",
    "    # Create file_name column to check\n",
    "    temp.insert(loc=0, column=\"file_name\", value=file_name)\n",
    "\n",
    "    # Create new columnn for data model name\n",
    "    temp.insert(\n",
    "        loc=1,\n",
    "        column=\"dm\",\n",
    "        value=re.sub(\n",
    "            \"\\s\\s+\",\n",
    "            \" \",\n",
    "            re.sub(\n",
    "                \"_\",\n",
    "                \" \",\n",
    "                re.sub(\n",
    "                    \"(EL)|(RFC)|(\\.xlsx)|([Aa]ssay)|([Dd]ata [Mm]odel)\", \"\", file_name\n",
    "                ),\n",
    "            ).strip(),\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    dm = pd.concat([dm, temp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 370 entries, 0 to 26\n",
      "Data columns (total 15 columns):\n",
      " #   Column                   Non-Null Count  Dtype  \n",
      "---  ------                   --------------  -----  \n",
      " 0   file_name                370 non-null    object \n",
      " 1   dm                       370 non-null    object \n",
      " 2   key                      370 non-null    object \n",
      " 3   description              370 non-null    object \n",
      " 4   valid values             221 non-null    object \n",
      " 5   required                 364 non-null    object \n",
      " 6   requires                 367 non-null    object \n",
      " 7   multivalue               365 non-null    object \n",
      " 8   type                     368 non-null    object \n",
      " 9   concept source ontology  367 non-null    object \n",
      " 10  note                     8 non-null      object \n",
      " 11  Unnamed: 9               0 non-null      float64\n",
      " 12  Unnamed: 10              2 non-null      object \n",
      " 13  ontology                 42 non-null     object \n",
      " 14  term id                  15 non-null     object \n",
      "dtypes: float64(1), object(14)\n",
      "memory usage: 46.2+ KB\n"
     ]
    }
   ],
   "source": [
    "dm.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>dm</th>\n",
       "      <th>key</th>\n",
       "      <th>description</th>\n",
       "      <th>valid values</th>\n",
       "      <th>required</th>\n",
       "      <th>requires</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>type</th>\n",
       "      <th>concept source ontology</th>\n",
       "      <th>note</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>ontology</th>\n",
       "      <th>term id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>specimenID</td>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td>n/a (unique to each data contributor)</td>\n",
       "      <td>True</td>\n",
       "      <td>scRNAseq, Biospecimen</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>sampleType</td>\n",
       "      <td>The type of sample collected</td>\n",
       "      <td>Amniotic Fluid,\\nAppendix,\\nB cell,\\nBasophils...</td>\n",
       "      <td>True</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nImmPort</td>\n",
       "      <td>The sample types are adopted from Uberon, Cell...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>scRNAseq,\\nsampleType = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,\\n1D Gel,\\n2D Gel,\\nA...</td>\n",
       "      <td>True</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>scRNAseq,\\nmeasurementTechnique = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            file_name        dm                          key  \\\n",
       "0  EL Assay_ scRNAseq data model.xlsx  scRNAseq                   specimenID   \n",
       "1  EL Assay_ scRNAseq data model.xlsx  scRNAseq                   sampleType   \n",
       "2  EL Assay_ scRNAseq data model.xlsx  scRNAseq            specifySampleType   \n",
       "3  EL Assay_ scRNAseq data model.xlsx  scRNAseq         measurementTechnique   \n",
       "4  EL Assay_ scRNAseq data model.xlsx  scRNAseq  specifyMeasurementTechnique   \n",
       "\n",
       "                                         description  \\\n",
       "0  Identifying string linked to a particular samp...   \n",
       "1                       The type of sample collected   \n",
       "2     If \"other\" is selected list the type of sample   \n",
       "3  The measurement technique describing the assay...   \n",
       "4  If \"other\" is selected list the name of the me...   \n",
       "\n",
       "                                        valid values required  \\\n",
       "0              n/a (unique to each data contributor)     True   \n",
       "1  Amniotic Fluid,\\nAppendix,\\nB cell,\\nBasophils...     True   \n",
       "2                                                NaN    False   \n",
       "3  16S rRNA gene sequencing,\\n1D Gel,\\n2D Gel,\\nA...     True   \n",
       "4                                                NaN    False   \n",
       "\n",
       "                                  requires multivalue    type  \\\n",
       "0                    scRNAseq, Biospecimen      False  string   \n",
       "1                                 scRNAseq      False  string   \n",
       "2            scRNAseq,\\nsampleType = other      False  string   \n",
       "3                                 scRNAseq      False  string   \n",
       "4  scRNAseq,\\nmeasurementTechnique = other      False  string   \n",
       "\n",
       "      concept source ontology  \\\n",
       "0            Sage Bionetworks   \n",
       "1  Sage Bionetworks,\\nImmPort   \n",
       "2            Sage Bionetworks   \n",
       "3            Sage Bionetworks   \n",
       "4            Sage Bionetworks   \n",
       "\n",
       "                                                note  Unnamed: 9  \\\n",
       "0                                                NaN         NaN   \n",
       "1  The sample types are adopted from Uberon, Cell...         NaN   \n",
       "2                                                NaN         NaN   \n",
       "3                                                NaN         NaN   \n",
       "4                                                NaN         NaN   \n",
       "\n",
       "        Unnamed: 10 ontology term id  \n",
       "0  Sage Bionetworks      NaN     NaN  \n",
       "1               NaN      NaN     NaN  \n",
       "2               NaN      NaN     NaN  \n",
       "3               NaN      NaN     NaN  \n",
       "4               NaN      NaN     NaN  "
      ]
     },
     "execution_count": 276,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# initial cleaning\n",
    "dm[[\"required\", \"multivalue\"]] = (\n",
    "    dm[[\"required\", \"multivalue\"]]\n",
    "    .fillna(False)\n",
    "    .astype(str)\n",
    "    .replace({\"1.0\": True, \"0.0\": False})\n",
    ")\n",
    "\n",
    "dm.fillna(\"\")\n",
    "dm.reset_index(drop=True, inplace=True)\n",
    "dm.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data model clean up\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/p0/5m4pdsm55jn_d5nzbjv6bjf40000gq/T/ipykernel_97139/1778036307.py:13: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  dm[\"valid values\"].str.contains(\n"
     ]
    }
   ],
   "source": [
    "# collapse presumed ontology columns and join with existing\n",
    "dm.loc[:, \"ontology\"] = dm.iloc[:, 11:].bfill(axis=1).iloc[:, 0]\n",
    "dm[\"ontology\"] = (\n",
    "    dm[[\"concept source ontology\", \"ontology\"]]\n",
    "    .fillna(\"\")\n",
    "    .apply(lambda x: \",\".join([y.strip() for y in x.unique() if len(y) > 0]), axis=1)\n",
    ")\n",
    "\n",
    "# if unique values are provided by data contributor then add this note in the ontology\n",
    "\n",
    "dm[\"ontology\"] = (\n",
    "    dm.loc[\n",
    "        dm[\"valid values\"].str.contains(\n",
    "            \"(n/a \\(unique to each data contributor\\))\", na=False\n",
    "        ),\n",
    "        \"ontology\",\n",
    "    ]\n",
    "    + \",\"\n",
    "    + \"Data Contributor\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>dm</th>\n",
       "      <th>key</th>\n",
       "      <th>description</th>\n",
       "      <th>valid values</th>\n",
       "      <th>required</th>\n",
       "      <th>requires</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>type</th>\n",
       "      <th>concept source ontology</th>\n",
       "      <th>note</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>ontology</th>\n",
       "      <th>term id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[specimenID]</td>\n",
       "      <td>[Identifying string linked to a particular sam...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[scRNAseq,  Biospecimen]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>[Sage Bionetworks, Data Contributor]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[sampleType]</td>\n",
       "      <td>[The type of sample collected]</td>\n",
       "      <td>[Amniotic Fluid, Appendix, B cell, Basophils, ...</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks, ImmPort]</td>\n",
       "      <td>[The sample types are adopted from Uberon,  Ce...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[specifySampleType]</td>\n",
       "      <td>[If \"other\" is selected list the type of sample]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[scRNAseq, sampleType = other]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[measurementTechnique]</td>\n",
       "      <td>[The measurement technique describing the assa...</td>\n",
       "      <td>[16S rRNA gene sequencing, 1D Gel, 2D Gel, Arr...</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[specifyMeasurementTechnique]</td>\n",
       "      <td>[If \"other\" is selected list the name of the m...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[scRNAseq, measurementTechnique = other]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              file_name          dm  \\\n",
       "0  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "1  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "2  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "3  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "4  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "\n",
       "                             key  \\\n",
       "0                   [specimenID]   \n",
       "1                   [sampleType]   \n",
       "2            [specifySampleType]   \n",
       "3         [measurementTechnique]   \n",
       "4  [specifyMeasurementTechnique]   \n",
       "\n",
       "                                         description  \\\n",
       "0  [Identifying string linked to a particular sam...   \n",
       "1                     [The type of sample collected]   \n",
       "2   [If \"other\" is selected list the type of sample]   \n",
       "3  [The measurement technique describing the assa...   \n",
       "4  [If \"other\" is selected list the name of the m...   \n",
       "\n",
       "                                        valid values required  \\\n",
       "0                                                 []   [True]   \n",
       "1  [Amniotic Fluid, Appendix, B cell, Basophils, ...   [True]   \n",
       "2                                                NaN  [False]   \n",
       "3  [16S rRNA gene sequencing, 1D Gel, 2D Gel, Arr...   [True]   \n",
       "4                                                NaN  [False]   \n",
       "\n",
       "                                   requires multivalue      type  \\\n",
       "0                  [scRNAseq,  Biospecimen]    [False]  [string]   \n",
       "1                                [scRNAseq]    [False]  [string]   \n",
       "2            [scRNAseq, sampleType = other]    [False]  [string]   \n",
       "3                                [scRNAseq]    [False]  [string]   \n",
       "4  [scRNAseq, measurementTechnique = other]    [False]  [string]   \n",
       "\n",
       "       concept source ontology  \\\n",
       "0           [Sage Bionetworks]   \n",
       "1  [Sage Bionetworks, ImmPort]   \n",
       "2           [Sage Bionetworks]   \n",
       "3           [Sage Bionetworks]   \n",
       "4           [Sage Bionetworks]   \n",
       "\n",
       "                                                note  Unnamed: 9  \\\n",
       "0                                                NaN         NaN   \n",
       "1  [The sample types are adopted from Uberon,  Ce...         NaN   \n",
       "2                                                NaN         NaN   \n",
       "3                                                NaN         NaN   \n",
       "4                                                NaN         NaN   \n",
       "\n",
       "          Unnamed: 10                              ontology term id  \n",
       "0  [Sage Bionetworks]  [Sage Bionetworks, Data Contributor]     NaN  \n",
       "1                 NaN                                   NaN     NaN  \n",
       "2                 NaN                                   NaN     NaN  \n",
       "3                 NaN                                   NaN     NaN  \n",
       "4                 NaN                                   NaN     NaN  "
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm = dm.apply(\n",
    "    lambda x: x.str.replace(\n",
    "        pat=\"\\n|(n/a \\(unique to each data contributor\\))\", repl=\"\", regex=True\n",
    "    ).str.split(\",\"),\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Parent</th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>specimenID</td>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq, Biospecimen</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,Data Contributor</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>sampleType</td>\n",
       "      <td>The type of sample collected</td>\n",
       "      <td>Amniotic Fluid,Appendix,B cell,Basophils,Bone,...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks,ImmPort</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq,sampleType = other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq,measurementTechnique = other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Parent                    Attribute  \\\n",
       "0  scRNAseq                   specimenID   \n",
       "1  scRNAseq                   sampleType   \n",
       "2  scRNAseq            specifySampleType   \n",
       "3  scRNAseq         measurementTechnique   \n",
       "4  scRNAseq  specifyMeasurementTechnique   \n",
       "\n",
       "                                         Description  \\\n",
       "0  Identifying string linked to a particular samp...   \n",
       "1                       The type of sample collected   \n",
       "2     If \"other\" is selected list the type of sample   \n",
       "3  The measurement technique describing the assay...   \n",
       "4  If \"other\" is selected list the name of the me...   \n",
       "\n",
       "                                        Valid Values Required DependsOn  \\\n",
       "0                                                        True             \n",
       "1  Amniotic Fluid,Appendix,B cell,Basophils,Bone,...     True             \n",
       "2                                                       False             \n",
       "3  16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...     True             \n",
       "4                                                       False             \n",
       "\n",
       "                     DependsOn Component Properties Validation Rules  \\\n",
       "0                  scRNAseq, Biospecimen                               \n",
       "1                               scRNAseq                               \n",
       "2            scRNAseq,sampleType = other                               \n",
       "3                               scRNAseq                               \n",
       "4  scRNAseq,measurementTechnique = other                               \n",
       "\n",
       "                     Source    Type                           Ontology  \\\n",
       "0          Sage Bionetworks  string  Sage Bionetworks,Data Contributor   \n",
       "1  Sage Bionetworks,ImmPort  string                                      \n",
       "2          Sage Bionetworks  string                                      \n",
       "3          Sage Bionetworks  string                                      \n",
       "4          Sage Bionetworks  string                                      \n",
       "\n",
       "  multivalue  \n",
       "0      False  \n",
       "1      False  \n",
       "2      False  \n",
       "3      False  \n",
       "4      False  "
      ]
     },
     "execution_count": 279,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# revert lists back to strings\n",
    "dm = dm.applymap(lambda x: join_strings(x))\n",
    "\n",
    "# Rename columns with DCA standards\n",
    "dm_schema_cols = {\n",
    "    \"dm\": \"Parent\",\n",
    "    \"key\": \"Attribute\",\n",
    "    \"description\": \"Description\",\n",
    "    \"valid values\": \"Valid Values\",\n",
    "    \"required\": \"Required\",\n",
    "    \"requires\": \"DependsOn Component\",\n",
    "    \"concept source ontology\": \"Source\",\n",
    "    \"ontology\": \"Ontology\",\n",
    "    \"type\": \"Type\",\n",
    "}\n",
    "\n",
    "dm = dm.rename(dm_schema_cols, axis=1)\n",
    "\n",
    "# drop unimportant columns\n",
    "r = re.compile(\"Unnamed*\", re.IGNORECASE)\n",
    "\n",
    "# Add additional required columns for DCA\n",
    "dm[\"Properties\"] = \"\"\n",
    "dm[\"Validation Rules\"] = dm[\"Type\"].map(validation_coder)\n",
    "dm[\"DependsOn\"] = \"\"\n",
    "# dm['DependsOn Component'] = \"\"\n",
    "\n",
    "keep_cols = [\n",
    "    \"Parent\",\n",
    "    \"Attribute\",\n",
    "    \"Description\",\n",
    "    \"Valid Values\",\n",
    "    \"Required\",\n",
    "    \"DependsOn\",\n",
    "    \"DependsOn Component\",\n",
    "    \"Properties\",\n",
    "    \"Validation Rules\",\n",
    "    \"Source\",\n",
    "    \"Type\",\n",
    "    \"Ontology\",\n",
    "    \"multivalue\",\n",
    "]\n",
    "\n",
    "dm = dm[keep_cols]\n",
    "\n",
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Parent</th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>proteomics</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>WGS</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>Microbiome</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>microbiome</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>genotyping</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>genotyping</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks,ImmPort</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>Biospecimen human</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>bsSeq</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>RNAseq</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>RNAseq</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Parent             Attribute  \\\n",
       "3                                            scRNAseq  measurementTechnique   \n",
       "56                                         proteomics  measurementTechnique   \n",
       "98                            Whole Genome Sequencing  measurementTechnique   \n",
       "130                                        Microbiome  measurementTechnique   \n",
       "195                                Metabolomics Human  measurementTechnique   \n",
       "279                                        genotyping  measurementTechnique   \n",
       "284                                 Biospecimen human  measurementTechnique   \n",
       "299  bsSeq (bisulfite-seq WGBS methylseq methylomics)  measurementTechnique   \n",
       "346                                            RNAseq  measurementTechnique   \n",
       "\n",
       "                                           Description  \\\n",
       "3    The measurement technique describing the assay...   \n",
       "56   The measurement technique describing the assay...   \n",
       "98   The measurement technique describing the assay...   \n",
       "130  The measurement technique describing the assay...   \n",
       "195  The measurement technique describing the assay...   \n",
       "279  The measurement technique describing the assay...   \n",
       "284  The measurement technique describing the assay...   \n",
       "299  The measurement technique describing the assay...   \n",
       "346  The measurement technique describing the assay...   \n",
       "\n",
       "                                          Valid Values Required DependsOn  \\\n",
       "3    16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...     True             \n",
       "56   16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...     True             \n",
       "98   16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...     True             \n",
       "130  16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...     True             \n",
       "195  16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...     True             \n",
       "279  16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...                      \n",
       "284  16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...     True             \n",
       "299  16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...     True             \n",
       "346  16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...     True             \n",
       "\n",
       "       DependsOn Component Properties Validation Rules  \\\n",
       "3                 scRNAseq                               \n",
       "56    mass spec proteomics                               \n",
       "98                     WGS                               \n",
       "130             microbiome                               \n",
       "195  mass spec metabolmics                               \n",
       "279            genotyping                                \n",
       "284            Biospecimen                               \n",
       "299                 bsSeq                                \n",
       "346                 RNAseq                               \n",
       "\n",
       "                       Source    Type Ontology multivalue  \n",
       "3            Sage Bionetworks  string               False  \n",
       "56           Sage Bionetworks  string               False  \n",
       "98           Sage Bionetworks  string               False  \n",
       "130          Sage Bionetworks  string               False  \n",
       "195          Sage Bionetworks  string                      \n",
       "279  Sage Bionetworks,ImmPort  string                      \n",
       "284          Sage Bionetworks  string               False  \n",
       "299          Sage Bionetworks  string               false  \n",
       "346          Sage Bionetworks  string               False  "
      ]
     },
     "execution_count": 280,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm.query('Attribute == \"measurementTechnique\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Parent</th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>acquisitionMode</td>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>acquisitionSoftware</td>\n",
       "      <td>The name of the acquisition software used</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>acquistionSoftwareVersion</td>\n",
       "      <td>The version number of software used</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>number</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>Biospecimen human</td>\n",
       "      <td>cellType</td>\n",
       "      <td>Indicate the cell type.</td>\n",
       "      <td>CCF-BSO values (not listed for purposes of thi...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>https://bioportal.bioontology.org/ontologies/C...</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>Individual nonHuman</td>\n",
       "      <td>cohort</td>\n",
       "      <td>Name of the cohort the individual belongs to</td>\n",
       "      <td>Other, Unknown,Not collected,Not applicable</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Individual</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>Microbiome</td>\n",
       "      <td>totalReads</td>\n",
       "      <td>Total number of sequencing reads from a library</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>microbiome</td>\n",
       "      <td></td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>number</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>transcriptType</td>\n",
       "      <td>The type of transcript reported</td>\n",
       "      <td>lincRNA,mRNA,snRNA,OtherUnknown,Not collected,...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>WGS</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ImmPort</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>RNAseq</td>\n",
       "      <td>transcriptType</td>\n",
       "      <td>The type of transcript reported</td>\n",
       "      <td>lincRNA,mRNA,snRNA,OtherUnknown,Not collected,...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>RNAseq</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ImmPort</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>RNAseq</td>\n",
       "      <td>valueReported</td>\n",
       "      <td>The count or gene count for the transcript</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>RNAseq</td>\n",
       "      <td></td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>number</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>valueReported</td>\n",
       "      <td>The count or gene count for the transcript</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>WGS</td>\n",
       "      <td></td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>number</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190 rows  13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Parent                  Attribute  \\\n",
       "215       Metabolomics Human            acquisitionMode   \n",
       "216       Metabolomics Human        acquisitionSoftware   \n",
       "217       Metabolomics Human  acquistionSoftwareVersion   \n",
       "293        Biospecimen human                   cellType   \n",
       "181      Individual nonHuman                     cohort   \n",
       "..                       ...                        ...   \n",
       "159               Microbiome                 totalReads   \n",
       "107  Whole Genome Sequencing             transcriptType   \n",
       "355                   RNAseq             transcriptType   \n",
       "358                   RNAseq              valueReported   \n",
       "110  Whole Genome Sequencing              valueReported   \n",
       "\n",
       "                                           Description  \\\n",
       "215  The specific aspect of a mass spectrometer met...   \n",
       "216          The name of the acquisition software used   \n",
       "217                The version number of software used   \n",
       "293                           Indicate the cell type.    \n",
       "181       Name of the cohort the individual belongs to   \n",
       "..                                                 ...   \n",
       "159    Total number of sequencing reads from a library   \n",
       "107                    The type of transcript reported   \n",
       "355                    The type of transcript reported   \n",
       "358         The count or gene count for the transcript   \n",
       "110         The count or gene count for the transcript   \n",
       "\n",
       "                                          Valid Values Required DependsOn  \\\n",
       "215  Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "216  Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "217                                                        True             \n",
       "293  CCF-BSO values (not listed for purposes of thi...     True             \n",
       "181        Other, Unknown,Not collected,Not applicable     True             \n",
       "..                                                 ...      ...       ...   \n",
       "159                                                        True             \n",
       "107  lincRNA,mRNA,snRNA,OtherUnknown,Not collected,...     True             \n",
       "355  lincRNA,mRNA,snRNA,OtherUnknown,Not collected,...     True             \n",
       "358                                                        True             \n",
       "110                                                        True             \n",
       "\n",
       "       DependsOn Component Properties  \\\n",
       "215  mass spec metabolmics              \n",
       "216  mass spec metabolmics              \n",
       "217  mass spec metabolmics              \n",
       "293            Biospecimen              \n",
       "181             Individual              \n",
       "..                     ...        ...   \n",
       "159             microbiome              \n",
       "107                    WGS              \n",
       "355                 RNAseq              \n",
       "358                 RNAseq              \n",
       "110                    WGS              \n",
       "\n",
       "                             Validation Rules  \\\n",
       "215                                             \n",
       "216                                             \n",
       "217  regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)   \n",
       "293                                             \n",
       "181                                             \n",
       "..                                        ...   \n",
       "159  regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)   \n",
       "107                                             \n",
       "355                                             \n",
       "358  regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)   \n",
       "110  regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)   \n",
       "\n",
       "                                                Source    Type Ontology  \\\n",
       "215  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...  string            \n",
       "216          http://purl.obolibrary.org/obo/MS_1001455  string            \n",
       "217                                   Sage Bionetworks  number            \n",
       "293  https://bioportal.bioontology.org/ontologies/C...  string            \n",
       "181                                   Sage Bionetworks  string            \n",
       "..                                                 ...     ...      ...   \n",
       "159                                   Sage Bionetworks  number            \n",
       "107                                            ImmPort  string            \n",
       "355                                            ImmPort  string            \n",
       "358                                   Sage Bionetworks  number            \n",
       "110                                   Sage Bionetworks  number            \n",
       "\n",
       "    multivalue  \n",
       "215             \n",
       "216             \n",
       "217             \n",
       "293      False  \n",
       "181       True  \n",
       "..         ...  \n",
       "159      False  \n",
       "107      False  \n",
       "355      False  \n",
       "358      False  \n",
       "110      False  \n",
       "\n",
       "[190 rows x 13 columns]"
      ]
     },
     "execution_count": 281,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm[dm.duplicated(subset=\"Attribute\")].sort_values(by=\"Attribute\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine duplicated attributes\n",
    "dm = dm.groupby(\"Attribute\").agg(lambda x: \",\".join(set(x.astype(str)))).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>acquisitionBatchID</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Acquisition batch identifier, provided by the ...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acquisitionBatchSize</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>The number of samples</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>acquisitionBatchSizeUnit</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>The unit of measurement for number of samples ...</td>\n",
       "      <td>AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics,acquisitionBatchSize</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>acquisitionMode</td>\n",
       "      <td>proteomics,Metabolomics Human</td>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec proteomics,mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>,False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>acquisitionSoftware</td>\n",
       "      <td>proteomics,Metabolomics Human</td>\n",
       "      <td>The name of the acquisition software used</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec proteomics,mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>,False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Attribute                         Parent  \\\n",
       "0        acquisitionBatchID             Metabolomics Human   \n",
       "1      acquisitionBatchSize             Metabolomics Human   \n",
       "2  acquisitionBatchSizeUnit             Metabolomics Human   \n",
       "3           acquisitionMode  proteomics,Metabolomics Human   \n",
       "4       acquisitionSoftware  proteomics,Metabolomics Human   \n",
       "\n",
       "                                         Description  \\\n",
       "0  Acquisition batch identifier, provided by the ...   \n",
       "1                              The number of samples   \n",
       "2  The unit of measurement for number of samples ...   \n",
       "3  The specific aspect of a mass spectrometer met...   \n",
       "4          The name of the acquisition software used   \n",
       "\n",
       "                                        Valid Values Required DependsOn  \\\n",
       "0                                                       False             \n",
       "1                                                       False             \n",
       "2  AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...    False             \n",
       "3  Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "4  Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "\n",
       "                          DependsOn Component Properties Validation Rules  \\\n",
       "0                       mass spec metabolmics                               \n",
       "1                       mass spec metabolmics                               \n",
       "2  mass spec metabolmics,acquisitionBatchSize                               \n",
       "3  mass spec proteomics,mass spec metabolmics                               \n",
       "4  mass spec proteomics,mass spec metabolmics                               \n",
       "\n",
       "                                              Source    Type Ontology  \\\n",
       "0                                   Sage Bionetworks  string            \n",
       "1                                   Sage Bionetworks  string            \n",
       "2                                   Sage Bionetworks  string            \n",
       "3  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...  string            \n",
       "4          http://purl.obolibrary.org/obo/MS_1001455  string            \n",
       "\n",
       "  multivalue  \n",
       "0             \n",
       "1             \n",
       "2             \n",
       "3     ,False  \n",
       "4     ,False  "
      ]
     },
     "execution_count": 283,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Dropping measurement technique\n",
    "# dm = dm.drop(\n",
    "#     index=dm.query('Attribute == \"measurementTechnique\"').index.values\n",
    "# ).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dm.reset_index(drop=True, inplace=True)\n",
    "\n",
    "# # QA Check\n",
    "# dm[\"Parent\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Metabolomics Human', 'proteomics,Metabolomics Human',\n",
       "       'Individual Human', 'Biospecimen nonHuman', 'Individual nonHuman',\n",
       "       'Biospecimen nonHuman,Biospecimen human', 'proteomics',\n",
       "       'Individual nonHuman,Individual Human',\n",
       "       'bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Microbiome,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Metabolomics Human,Microbiome',\n",
       "       'Individual nonHuman,Biospecimen nonHuman,Individual Human,Biospecimen human',\n",
       "       'Microbiome,Whole Genome Sequencing,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),RNAseq',\n",
       "       'Whole Genome Sequencing,scRNAseq,Microbiome,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Metabolomics Human,proteomics,Microbiome,Whole Genome Sequencing,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),RNAseq,Biospecimen human,genotyping',\n",
       "       'Metabolomics Human,proteomics,Microbiome,Whole Genome Sequencing,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),RNAseq',\n",
       "       'genotyping',\n",
       "       'Whole Genome Sequencing,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),RNAseq',\n",
       "       'Whole Genome Sequencing,scRNAseq,RNAseq',\n",
       "       'Whole Genome Sequencing,scRNAseq', 'scRNAseq',\n",
       "       'Metabolomics Human,proteomics,Microbiome,Whole Genome Sequencing,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),RNAseq,genotyping',\n",
       "       'scRNAseq,Microbiome,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Microbiome',\n",
       "       'Metabolomics Human,proteomics,Biospecimen nonHuman,Microbiome,Whole Genome Sequencing,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),RNAseq,Biospecimen human,genotyping'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 286,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# QA Check\n",
    "dm[\"Parent\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Attribute              object\n",
       "Parent                 object\n",
       "Description            object\n",
       "Valid Values           object\n",
       "Required               object\n",
       "DependsOn              object\n",
       "DependsOn Component    object\n",
       "Properties             object\n",
       "Validation Rules       object\n",
       "Source                 object\n",
       "Type                   object\n",
       "Ontology               object\n",
       "multivalue             object\n",
       "dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>acquisitionBatchID</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Acquisition batch identifier, provided by the ...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acquisitionBatchSize</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>The number of samples</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>acquisitionBatchSizeUnit</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>The unit of measurement for number of samples ...</td>\n",
       "      <td>AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics,acquisitionBatchSize</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>acquisitionMode</td>\n",
       "      <td>proteomics,Metabolomics Human</td>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec proteomics,mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>,False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>acquisitionSoftware</td>\n",
       "      <td>proteomics,Metabolomics Human</td>\n",
       "      <td>The name of the acquisition software used</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec proteomics,mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>,False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Attribute                         Parent  \\\n",
       "0        acquisitionBatchID             Metabolomics Human   \n",
       "1      acquisitionBatchSize             Metabolomics Human   \n",
       "2  acquisitionBatchSizeUnit             Metabolomics Human   \n",
       "3           acquisitionMode  proteomics,Metabolomics Human   \n",
       "4       acquisitionSoftware  proteomics,Metabolomics Human   \n",
       "\n",
       "                                         Description  \\\n",
       "0  Acquisition batch identifier, provided by the ...   \n",
       "1                              The number of samples   \n",
       "2  The unit of measurement for number of samples ...   \n",
       "3  The specific aspect of a mass spectrometer met...   \n",
       "4          The name of the acquisition software used   \n",
       "\n",
       "                                        Valid Values Required DependsOn  \\\n",
       "0                                                       False             \n",
       "1                                                       False             \n",
       "2  AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...    False             \n",
       "3  Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "4  Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "\n",
       "                          DependsOn Component Properties Validation Rules  \\\n",
       "0                       mass spec metabolmics                               \n",
       "1                       mass spec metabolmics                               \n",
       "2  mass spec metabolmics,acquisitionBatchSize                               \n",
       "3  mass spec proteomics,mass spec metabolmics                               \n",
       "4  mass spec proteomics,mass spec metabolmics                               \n",
       "\n",
       "                                              Source    Type Ontology  \\\n",
       "0                                   Sage Bionetworks  string            \n",
       "1                                   Sage Bionetworks  string            \n",
       "2                                   Sage Bionetworks  string            \n",
       "3  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...  string            \n",
       "4          http://purl.obolibrary.org/obo/MS_1001455  string            \n",
       "\n",
       "  multivalue  \n",
       "0             \n",
       "1             \n",
       "2             \n",
       "3     ,False  \n",
       "4     ,False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "array(['Metabolomics Human', 'proteomics,Metabolomics Human',\n",
       "       'Individual Human', 'Biospecimen nonHuman', 'Individual nonHuman',\n",
       "       'Biospecimen nonHuman,Biospecimen human', 'proteomics',\n",
       "       'Individual nonHuman,Individual Human',\n",
       "       'bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Microbiome,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Metabolomics Human,Microbiome',\n",
       "       'Individual nonHuman,Biospecimen nonHuman,Individual Human,Biospecimen human',\n",
       "       'Microbiome,Whole Genome Sequencing,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),RNAseq',\n",
       "       'Whole Genome Sequencing,scRNAseq,Microbiome,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Metabolomics Human,proteomics,Microbiome,Whole Genome Sequencing,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),RNAseq,Biospecimen human,genotyping',\n",
       "       'Metabolomics Human,proteomics,Microbiome,Whole Genome Sequencing,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),RNAseq',\n",
       "       'genotyping',\n",
       "       'Whole Genome Sequencing,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),RNAseq',\n",
       "       'Whole Genome Sequencing,scRNAseq,RNAseq',\n",
       "       'Whole Genome Sequencing,scRNAseq', 'scRNAseq',\n",
       "       'Metabolomics Human,proteomics,Microbiome,Whole Genome Sequencing,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),RNAseq,genotyping',\n",
       "       'scRNAseq,Microbiome,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Microbiome',\n",
       "       'Metabolomics Human,proteomics,Biospecimen nonHuman,Microbiome,Whole Genome Sequencing,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),RNAseq,Biospecimen human,genotyping'],\n",
       "      dtype=object)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(dm.dtypes)\n",
    "display(dm.head())\n",
    "display(dm.Parent.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Attribute',\n",
       " 'DependsOn',\n",
       " 'DependsOn Component',\n",
       " 'Description',\n",
       " 'Ontology',\n",
       " 'Parent',\n",
       " 'Properties',\n",
       " 'Required',\n",
       " 'Source',\n",
       " 'Type',\n",
       " 'Valid Values',\n",
       " 'Validation Rules',\n",
       " 'multivalue'}"
      ]
     },
     "execution_count": 288,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " set(dm.columns.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Attribute',\n",
       " 'DependsOn',\n",
       " 'DependsOn Component',\n",
       " 'Description',\n",
       " 'Ontology',\n",
       " 'Parent',\n",
       " 'Properties',\n",
       " 'Required',\n",
       " 'Source',\n",
       " 'Type',\n",
       " 'Valid Values',\n",
       " 'Validation Rules',\n",
       " 'multivalue'}"
      ]
     },
     "execution_count": 289,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = set(base_cols)\n",
    "test.update(set(dm.columns.tolist()))\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Attribute',\n",
       " 'DependsOn',\n",
       " 'DependsOn Component',\n",
       " 'Description',\n",
       " 'Ontology',\n",
       " 'Parent',\n",
       " 'Properties',\n",
       " 'Required',\n",
       " 'Source',\n",
       " 'Type',\n",
       " 'Valid Values',\n",
       " 'Validation Rules',\n",
       " 'multivalue'}"
      ]
     },
     "execution_count": 290,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reorder Columns based on DCA Standards\n",
    "dm = dm.loc[:, base_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>acquisitionBatchID</td>\n",
       "      <td>Acquisition batch identifier, provided by the ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acquisitionBatchSize</td>\n",
       "      <td>The number of samples</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>acquisitionBatchSizeUnit</td>\n",
       "      <td>The unit of measurement for number of samples ...</td>\n",
       "      <td>AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>mass spec metabolmics,acquisitionBatchSize</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>acquisitionMode</td>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,Metabolomics Human</td>\n",
       "      <td>mass spec proteomics,mass spec metabolmics</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>acquisitionSoftware</td>\n",
       "      <td>The name of the acquisition software used</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,Metabolomics Human</td>\n",
       "      <td>mass spec proteomics,mass spec metabolmics</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>useTreatment?</td>\n",
       "      <td>Was a treatment applied to the sample?</td>\n",
       "      <td>Yes,No</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>genotyping</td>\n",
       "      <td>genotyping</td>\n",
       "      <td>Sage Bionetworks,ImmPort</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>vacuumPressure</td>\n",
       "      <td>The recorded vacuum pressure value</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>mass spec metabolmics,hasIonizationSource? = Yes</td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>vacuumPressureUnit</td>\n",
       "      <td>Unit of vacuum pressure value</td>\n",
       "      <td>AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>mass spec metabolmics,hasIonizationSource? = Y...</td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>valueReported</td>\n",
       "      <td>The count or gene count for the transcript</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Whole Genome Sequencing,scRNAseq,RNAseq</td>\n",
       "      <td>WGS,scRNAseq,RNAseq</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>visitCode</td>\n",
       "      <td>Indicate which longitudinal visit for the indi...</td>\n",
       "      <td>,Other,Unknown,Not collected,Not applicable</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Individual Human</td>\n",
       "      <td>Individual</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>180 rows  10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Attribute  \\\n",
       "0          acquisitionBatchID   \n",
       "1        acquisitionBatchSize   \n",
       "2    acquisitionBatchSizeUnit   \n",
       "3             acquisitionMode   \n",
       "4         acquisitionSoftware   \n",
       "..                        ...   \n",
       "175             useTreatment?   \n",
       "176            vacuumPressure   \n",
       "177        vacuumPressureUnit   \n",
       "178             valueReported   \n",
       "179                 visitCode   \n",
       "\n",
       "                                           Description  \\\n",
       "0    Acquisition batch identifier, provided by the ...   \n",
       "1                                The number of samples   \n",
       "2    The unit of measurement for number of samples ...   \n",
       "3    The specific aspect of a mass spectrometer met...   \n",
       "4            The name of the acquisition software used   \n",
       "..                                                 ...   \n",
       "175             Was a treatment applied to the sample?   \n",
       "176                 The recorded vacuum pressure value   \n",
       "177                      Unit of vacuum pressure value   \n",
       "178         The count or gene count for the transcript   \n",
       "179  Indicate which longitudinal visit for the indi...   \n",
       "\n",
       "                                          Valid Values DependsOn Properties  \\\n",
       "0                                                                             \n",
       "1                                                                             \n",
       "2    AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...                        \n",
       "3    Unknown,Not collected,Not applicable,Not speci...                        \n",
       "4    Unknown,Not collected,Not applicable,Not speci...                        \n",
       "..                                                 ...       ...        ...   \n",
       "175                                             Yes,No                        \n",
       "176                                                                           \n",
       "177  AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...                        \n",
       "178                                                                           \n",
       "179        ,Other,Unknown,Not collected,Not applicable                        \n",
       "\n",
       "    Required                                   Parent  \\\n",
       "0      False                       Metabolomics Human   \n",
       "1      False                       Metabolomics Human   \n",
       "2      False                       Metabolomics Human   \n",
       "3       True            proteomics,Metabolomics Human   \n",
       "4       True            proteomics,Metabolomics Human   \n",
       "..       ...                                      ...   \n",
       "175                                        genotyping   \n",
       "176    False                       Metabolomics Human   \n",
       "177    False                       Metabolomics Human   \n",
       "178     True  Whole Genome Sequencing,scRNAseq,RNAseq   \n",
       "179     True                         Individual Human   \n",
       "\n",
       "                                   DependsOn Component  \\\n",
       "0                                mass spec metabolmics   \n",
       "1                                mass spec metabolmics   \n",
       "2           mass spec metabolmics,acquisitionBatchSize   \n",
       "3           mass spec proteomics,mass spec metabolmics   \n",
       "4           mass spec proteomics,mass spec metabolmics   \n",
       "..                                                 ...   \n",
       "175                                        genotyping    \n",
       "176   mass spec metabolmics,hasIonizationSource? = Yes   \n",
       "177  mass spec metabolmics,hasIonizationSource? = Y...   \n",
       "178                                WGS,scRNAseq,RNAseq   \n",
       "179                                         Individual   \n",
       "\n",
       "                                                Source  \\\n",
       "0                                     Sage Bionetworks   \n",
       "1                                     Sage Bionetworks   \n",
       "2                                     Sage Bionetworks   \n",
       "3    https://www.ebi.ac.uk/ols/ontologies/ms/terms?...   \n",
       "4            http://purl.obolibrary.org/obo/MS_1001455   \n",
       "..                                                 ...   \n",
       "175                           Sage Bionetworks,ImmPort   \n",
       "176  Sage Bionetworks,Proposed minimum metadata rel...   \n",
       "177  Sage Bionetworks,Proposed minimum metadata rel...   \n",
       "178                                   Sage Bionetworks   \n",
       "179                                   Sage Bionetworks   \n",
       "\n",
       "                             Validation Rules  \n",
       "0                                              \n",
       "1                                              \n",
       "2                                              \n",
       "3                                              \n",
       "4                                              \n",
       "..                                        ...  \n",
       "175                                       nan  \n",
       "176  regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)  \n",
       "177                                            \n",
       "178  regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)  \n",
       "179  regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)  \n",
       "\n",
       "[180 rows x 10 columns]"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean up\n",
    "\n",
    "- Cleanup `DependsOn Component`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "recoder = {\n",
    "    \"metabolmics\": \"metabolomics\",\n",
    "    \"(mass spec proteomics)\": \"Proteomics\",\n",
    "    \"(mass spec metabolomics)\": \"Metabolomics Human\",\n",
    "    \"(assay_otheruseTreatment? = Yes)\": \"assay_other, useTreatment? = Yes\",\n",
    "    \"OtherUnknown\": \"Other, Unknown\",\n",
    "    \"falseFalseFALSEtrueTrueTRUE\": \"TRUE, FALSE\",\n",
    "    re.compile(\"Forwardreverse\", flags=re.IGNORECASE): \"forward,reverse\",\n",
    "    re.compile(\"singleEndpairedEnd\"): \"singleEnd, pairedEnd\",\n",
    "    re.compile(\"(WGS)\"): \"Whole Genome Sequencing\",\n",
    "    re.compile(\"\\?\"): \"\",\n",
    "    \"Zeiss LSM 980Other\": \"Zeiss LSM 980\",\n",
    "    \"bsSeqsampleType = other\": \"bsSeq, sampleType = other\",\n",
    "    re.compile(\n",
    "        \"HPO, MONDO, MAXO codes or labels (not listed for purposes of this RFC)\"\n",
    "    ): \"HPO and MONDO and MAXO codes or labels (not listed for purposes of this RFC)\",\n",
    "}\n",
    "\n",
    "# 'mass spec metabolomics,measurementTechnique = other'\n",
    "# falseFalseFALSEtrueTrueTRUE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>cohort</td>\n",
       "      <td>Name of the cohort the individual belongs to</td>\n",
       "      <td>Centenarian, US Family, Denmark Family, The Osteoporotic Fractures in Men (MrOS) Study, Study of Osteoporotic Fractures (SOF), The Health, Aging, and Body Composition Study (HealthABC), Cardiovascular Health Study (CHS),Other,Unknown,Not collected,Not applicable</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Individual Human</td>\n",
       "      <td>Individual</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Attribute                                   Description  \\\n",
       "335    cohort  Name of the cohort the individual belongs to   \n",
       "\n",
       "                                                                                                                                                                                                                                                               Valid Values  \\\n",
       "335  Centenarian, US Family, Denmark Family, The Osteoporotic Fractures in Men (MrOS) Study, Study of Osteoporotic Fractures (SOF), The Health, Aging, and Body Composition Study (HealthABC), Cardiovascular Health Study (CHS),Other,Unknown,Not collected,Not applicable   \n",
       "\n",
       "    DependsOn Properties Required            Parent DependsOn Component  \\\n",
       "335                          True  Individual Human          Individual   \n",
       "\n",
       "               Source Validation Rules  \n",
       "335  Sage Bionetworks                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm[dm[\"Valid Values\"].str.contains(\"and Body Composition Study\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>ethnicity</td>\n",
       "      <td>Ethnicity of individual</td>\n",
       "      <td>Not Hispanic or Latino, Hispanic or Latino, Prefer not to answer,Other,Unknown,Not collected,Not applicable</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Individual Human</td>\n",
       "      <td>Individual</td>\n",
       "      <td>Sage Bionetworks,https://www.synapse.org/#!Synapse:syn25878249</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Attribute              Description  \\\n",
       "344  ethnicity  Ethnicity of individual   \n",
       "\n",
       "                                                                                                    Valid Values  \\\n",
       "344  Not Hispanic or Latino, Hispanic or Latino, Prefer not to answer,Other,Unknown,Not collected,Not applicable   \n",
       "\n",
       "    DependsOn Properties Required            Parent DependsOn Component  \\\n",
       "344                          True  Individual Human          Individual   \n",
       "\n",
       "                                                             Source  \\\n",
       "344  Sage Bionetworks,https://www.synapse.org/#!Synapse:syn25878249   \n",
       "\n",
       "    Validation Rules  \n",
       "344                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm.query('Attribute == \"ethnicity\"'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inconsistency with parent and depends on component\n",
    "dm[\"DependsOn Component\"] = dm[\"DependsOn Component\"].replace(recoder, regex=True)\n",
    "\n",
    "dm[\"Valid Values\"] = dm[\"Valid Values\"].replace(recoder, regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm = dm.apply(lambda x: x.replace(recoder, regex=True), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>specimenID</td>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Biospecimen,Metabolomics Human</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>sampleType</td>\n",
       "      <td>The type of sample collected</td>\n",
       "      <td>Amniotic Fluid,Appendix,B cell,Basophils,Bone,...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Sage Bionetworks,ImmPort</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Metabolomics Human,sampleType = other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Metabolomics HumanmeasurementTechnique = other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>technologyPlatformVersion</td>\n",
       "      <td>The specific version (application, manufacture...</td>\n",
       "      <td>Other, Unknown,Not collected,Not applicable,No...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C45378</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>processingBatchID</td>\n",
       "      <td>Processing batch identifier, provided by the d...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>processingBatchSize</td>\n",
       "      <td>The number of samples</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>processingBatchSizeUnit</td>\n",
       "      <td>The unit of measurement for number of samples ...</td>\n",
       "      <td>AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Metabolomics Human,processingBatchSize</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>specifyProcessingBatchSizeUnit</td>\n",
       "      <td>If \"other\" list units of measure</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Metabolomics Human,ProcessingBatchSizeUnit = o...</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>f</td>\n",
       "      <td>The data file provides the name of the raw fil...</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65 rows  10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Attribute  \\\n",
       "161                      specimenID   \n",
       "162                      sampleType   \n",
       "163               specifySampleType   \n",
       "164     specifyMeasurementTechnique   \n",
       "165       technologyPlatformVersion   \n",
       "..                              ...   \n",
       "222               processingBatchID   \n",
       "223             processingBatchSize   \n",
       "224         processingBatchSizeUnit   \n",
       "225  specifyProcessingBatchSizeUnit   \n",
       "226                               f   \n",
       "\n",
       "                                           Description  \\\n",
       "161  Identifying string linked to a particular samp...   \n",
       "162                       The type of sample collected   \n",
       "163     If \"other\" is selected list the type of sample   \n",
       "164  If \"other\" is selected list the name of the me...   \n",
       "165  The specific version (application, manufacture...   \n",
       "..                                                 ...   \n",
       "222  Processing batch identifier, provided by the d...   \n",
       "223                              The number of samples   \n",
       "224  The unit of measurement for number of samples ...   \n",
       "225                   If \"other\" list units of measure   \n",
       "226  The data file provides the name of the raw fil...   \n",
       "\n",
       "                                          Valid Values DependsOn Properties  \\\n",
       "161                                                                           \n",
       "162  Amniotic Fluid,Appendix,B cell,Basophils,Bone,...                        \n",
       "163                                                                           \n",
       "164                                                                           \n",
       "165  Other, Unknown,Not collected,Not applicable,No...                        \n",
       "..                                                 ...       ...        ...   \n",
       "222                                                                           \n",
       "223                                                                           \n",
       "224  AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...                        \n",
       "225                                                                           \n",
       "226  Unknown,Not collected,Not applicable,Not speci...                        \n",
       "\n",
       "    Required              Parent  \\\n",
       "161     True  Metabolomics Human   \n",
       "162     True  Metabolomics Human   \n",
       "163    False  Metabolomics Human   \n",
       "164    False  Metabolomics Human   \n",
       "165     True  Metabolomics Human   \n",
       "..       ...                 ...   \n",
       "222    False  Metabolomics Human   \n",
       "223    False  Metabolomics Human   \n",
       "224    False  Metabolomics Human   \n",
       "225    False  Metabolomics Human   \n",
       "226     True  Metabolomics Human   \n",
       "\n",
       "                                   DependsOn Component  \\\n",
       "161                     Biospecimen,Metabolomics Human   \n",
       "162                                 Metabolomics Human   \n",
       "163              Metabolomics Human,sampleType = other   \n",
       "164     Metabolomics HumanmeasurementTechnique = other   \n",
       "165                                 Metabolomics Human   \n",
       "..                                                 ...   \n",
       "222                                 Metabolomics Human   \n",
       "223                                 Metabolomics Human   \n",
       "224             Metabolomics Human,processingBatchSize   \n",
       "225  Metabolomics Human,ProcessingBatchSizeUnit = o...   \n",
       "226                                 Metabolomics Human   \n",
       "\n",
       "                                         Source Validation Rules  \n",
       "161                            Sage Bionetworks                   \n",
       "162                    Sage Bionetworks,ImmPort                   \n",
       "163                            Sage Bionetworks                   \n",
       "164                            Sage Bionetworks                   \n",
       "165  http://purl.obolibrary.org/obo/NCIT_C45378                   \n",
       "..                                          ...              ...  \n",
       "222                            Sage Bionetworks                   \n",
       "223                            Sage Bionetworks                   \n",
       "224                            Sage Bionetworks                   \n",
       "225                            Sage Bionetworks                   \n",
       "226                  Proteomics Sample Metadata                   \n",
       "\n",
       "[65 rows x 10 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# QA check\n",
    "dm[dm[\"DependsOn Component\"].str.contains(\"metabolomics\", case=False, na=False)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cleaning other values and equal values\n",
    "\n",
    "Removing illegal characters\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove any special characters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['reagentID(s)', 'treatmentID(s)']"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "old_values = (\n",
    "    dm[dm[\"Attribute\"].astype(str).str.contains(\"\\(|\\)|\\?\", na=False)][\"Attribute\"]\n",
    "    .unique()\n",
    "    .tolist()\n",
    ")\n",
    "\n",
    "old_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['reagentIDs', 'treatmentIDs']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_values = [re.sub(\"\\(|\\)|\\?\", \"\", t) for t in old_values]\n",
    "new_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{re.compile(r'reagentID\\(s\\)', re.UNICODE): 'reagentIDs',\n",
       " re.compile(r'treatmentID\\(s\\)', re.UNICODE): 'treatmentIDs'}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dictionary of replacement values for data model\n",
    "old_values = [re.compile(re.escape(ov)) for ov in old_values]\n",
    "\n",
    "original_recoder = dict(zip(old_values, new_values))\n",
    "\n",
    "original_recoder"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clean up equals in depends on\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split list to process other values\n",
    "# Find the other columns in the data model\n",
    "others = dm[dm[\"DependsOn Component\"].str.contains(\"=\", na=False)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "others[\"DependsOn Component Original\"] = others[\"DependsOn Component\"].str.split(\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>DependsOn Component Original</th>\n",
       "      <th>equals_series</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td>bsSeq, sampleType = other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>[bsSeq,  sampleType = other]</td>\n",
       "      <td>sampleType = other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td>bsSeq,measurementTechnique = other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>[bsSeq, measurementTechnique = other]</td>\n",
       "      <td>measurementTechnique = other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>specifyPlatformVersion</td>\n",
       "      <td>If \"other\" list the name of the platform version</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td>bsSeq,technologyPlatformVersion = other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>[bsSeq, technologyPlatformVersion = other]</td>\n",
       "      <td>technologyPlatformVersion = other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>specifyPlatformLocation</td>\n",
       "      <td>If \"other\" list the name of the platform location</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td>bsSeq,platformLocation = other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>[bsSeq, platformLocation = other]</td>\n",
       "      <td>platformLocation = other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>specifyDNABatchSizeUnit</td>\n",
       "      <td>If \"other\" list unit of measure</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td>bsSeq,dnaBatchSizeUnit = other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>[bsSeq, dnaBatchSizeUnit = other]</td>\n",
       "      <td>dnaBatchSizeUnit = other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>specifyOtherModificationParameters</td>\n",
       "      <td>If \"other\" is selected, list the custom modifi...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>Proteomics,modificationParameters = other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>[Proteomics, modificationParameters = other]</td>\n",
       "      <td>modificationParameters = other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>specifyOtherCleavageAgent</td>\n",
       "      <td>If \"other\" is selected, list the enzyme name.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>Proteomics,cleavageAgents = other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>[Proteomics, cleavageAgents = other]</td>\n",
       "      <td>cleavageAgents = other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>ethnicGroupCode</td>\n",
       "      <td>A coded value specifying the self-declared eth...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Individual Human</td>\n",
       "      <td>Individual,ethnicity = Hispanic or Latino</td>\n",
       "      <td>https://ncithesaurus.nci.nih.gov/ncitbrowser/p...</td>\n",
       "      <td></td>\n",
       "      <td>[Individual, ethnicity = Hispanic or Latino]</td>\n",
       "      <td>ethnicity = Hispanic or Latino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>diagnosis</td>\n",
       "      <td>Indicate the disease or condition.</td>\n",
       "      <td>HPO, MONDO, MAXO codes or labels (not listed f...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Individual Human</td>\n",
       "      <td>Individual,diagnosisStatus = true</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/mondo,htt...</td>\n",
       "      <td></td>\n",
       "      <td>[Individual, diagnosisStatus = true]</td>\n",
       "      <td>diagnosisStatus = true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>captivityDuration</td>\n",
       "      <td>The duration of captivity in months for the in...</td>\n",
       "      <td>Unknown,Not collected,Not applicable</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Individual nonHuman</td>\n",
       "      <td>Individual,captivityStatus = captive</td>\n",
       "      <td>https://www.animalaudiograms.org/audiogram_met...</td>\n",
       "      <td></td>\n",
       "      <td>[Individual, captivityStatus = captive]</td>\n",
       "      <td>captivityStatus = captive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117 rows  12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Attribute  \\\n",
       "2                     specifySampleType   \n",
       "3           specifyMeasurementTechnique   \n",
       "5                specifyPlatformVersion   \n",
       "7               specifyPlatformLocation   \n",
       "13              specifyDNABatchSizeUnit   \n",
       "..                                  ...   \n",
       "294  specifyOtherModificationParameters   \n",
       "296           specifyOtherCleavageAgent   \n",
       "345                     ethnicGroupCode   \n",
       "348                           diagnosis   \n",
       "360                   captivityDuration   \n",
       "\n",
       "                                           Description  \\\n",
       "2       If \"other\" is selected list the type of sample   \n",
       "3    If \"other\" is selected list the name of the me...   \n",
       "5     If \"other\" list the name of the platform version   \n",
       "7    If \"other\" list the name of the platform location   \n",
       "13                     If \"other\" list unit of measure   \n",
       "..                                                 ...   \n",
       "294  If \"other\" is selected, list the custom modifi...   \n",
       "296     If \"other\" is selected, list the enzyme name.    \n",
       "345  A coded value specifying the self-declared eth...   \n",
       "348                 Indicate the disease or condition.   \n",
       "360  The duration of captivity in months for the in...   \n",
       "\n",
       "                                          Valid Values DependsOn Properties  \\\n",
       "2                                                                             \n",
       "3                                                                             \n",
       "5                                                                             \n",
       "7                                                                             \n",
       "13                                                                            \n",
       "..                                                 ...       ...        ...   \n",
       "294                                                                           \n",
       "296                                                                           \n",
       "345                                                                           \n",
       "348  HPO, MONDO, MAXO codes or labels (not listed f...                        \n",
       "360               Unknown,Not collected,Not applicable                        \n",
       "\n",
       "    Required                                            Parent  \\\n",
       "2      False  bsSeq (bisulfite-seq WGBS methylseq methylomics)   \n",
       "3      False  bsSeq (bisulfite-seq WGBS methylseq methylomics)   \n",
       "5      False  bsSeq (bisulfite-seq WGBS methylseq methylomics)   \n",
       "7      False  bsSeq (bisulfite-seq WGBS methylseq methylomics)   \n",
       "13     False  bsSeq (bisulfite-seq WGBS methylseq methylomics)   \n",
       "..       ...                                               ...   \n",
       "294    False                                        proteomics   \n",
       "296    False                                        proteomics   \n",
       "345    False                                  Individual Human   \n",
       "348    False                                  Individual Human   \n",
       "360    False                               Individual nonHuman   \n",
       "\n",
       "                           DependsOn Component  \\\n",
       "2                    bsSeq, sampleType = other   \n",
       "3           bsSeq,measurementTechnique = other   \n",
       "5      bsSeq,technologyPlatformVersion = other   \n",
       "7               bsSeq,platformLocation = other   \n",
       "13              bsSeq,dnaBatchSizeUnit = other   \n",
       "..                                         ...   \n",
       "294  Proteomics,modificationParameters = other   \n",
       "296          Proteomics,cleavageAgents = other   \n",
       "345  Individual,ethnicity = Hispanic or Latino   \n",
       "348          Individual,diagnosisStatus = true   \n",
       "360       Individual,captivityStatus = captive   \n",
       "\n",
       "                                                Source Validation Rules  \\\n",
       "2                                     Sage Bionetworks                    \n",
       "3                                     Sage Bionetworks                    \n",
       "5                                     Sage Bionetworks                    \n",
       "7                                     Sage Bionetworks                    \n",
       "13                                    Sage Bionetworks                    \n",
       "..                                                 ...              ...   \n",
       "294                                   Sage Bionetworks                    \n",
       "296                                   Sage Bionetworks                    \n",
       "345  https://ncithesaurus.nci.nih.gov/ncitbrowser/p...                    \n",
       "348  https://www.ebi.ac.uk/ols/ontologies/mondo,htt...                    \n",
       "360  https://www.animalaudiograms.org/audiogram_met...                    \n",
       "\n",
       "                     DependsOn Component Original  \\\n",
       "2                    [bsSeq,  sampleType = other]   \n",
       "3           [bsSeq, measurementTechnique = other]   \n",
       "5      [bsSeq, technologyPlatformVersion = other]   \n",
       "7               [bsSeq, platformLocation = other]   \n",
       "13              [bsSeq, dnaBatchSizeUnit = other]   \n",
       "..                                            ...   \n",
       "294  [Proteomics, modificationParameters = other]   \n",
       "296          [Proteomics, cleavageAgents = other]   \n",
       "345  [Individual, ethnicity = Hispanic or Latino]   \n",
       "348          [Individual, diagnosisStatus = true]   \n",
       "360       [Individual, captivityStatus = captive]   \n",
       "\n",
       "                         equals_series  \n",
       "2                   sampleType = other  \n",
       "3         measurementTechnique = other  \n",
       "5    technologyPlatformVersion = other  \n",
       "7             platformLocation = other  \n",
       "13            dnaBatchSizeUnit = other  \n",
       "..                                 ...  \n",
       "294     modificationParameters = other  \n",
       "296             cleavageAgents = other  \n",
       "345     ethnicity = Hispanic or Latino  \n",
       "348             diagnosisStatus = true  \n",
       "360          captivityStatus = captive  \n",
       "\n",
       "[117 rows x 12 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create series of equals values to use for new attributes/ valid values relationship\n",
    "others[\"equals_series\"] = others[\"DependsOn Component Original\"].apply(\n",
    "    lambda x: [y for y in x if bool(re.search(\"=\", y))][0]\n",
    ")\n",
    "\n",
    "others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_new_value(old_value):\n",
    "    new_vals = old_value.split(\"=\")\n",
    "    new_vals = [nv.strip() for nv in new_vals]\n",
    "    # convert to camel case\n",
    "    nv = new_vals[1].capitalize() + new_vals[0][0].upper() + new_vals[0][1:]\n",
    "    return nv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "others[\"equals_attribute\"] = others[\"equals_series\"].apply(create_new_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "others[\"DependsOn Component\"] = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "equals_df = others.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "equals_df = equals_df.drop(columns=\"DependsOn\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OtherSampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td></td>\n",
       "      <td>specifySampleType</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OtherMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OtherTechnologyPlatformVersion</td>\n",
       "      <td>If \"other\" list the name of the platform version</td>\n",
       "      <td></td>\n",
       "      <td>specifyPlatformVersion</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>OtherPlatformLocation</td>\n",
       "      <td>If \"other\" list the name of the platform location</td>\n",
       "      <td></td>\n",
       "      <td>specifyPlatformLocation</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>OtherDnaBatchSizeUnit</td>\n",
       "      <td>If \"other\" list unit of measure</td>\n",
       "      <td></td>\n",
       "      <td>specifyDNABatchSizeUnit</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>OtherModificationParameters</td>\n",
       "      <td>If \"other\" is selected, list the custom modifi...</td>\n",
       "      <td></td>\n",
       "      <td>specifyOtherModificationParameters</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>OtherCleavageAgents</td>\n",
       "      <td>If \"other\" is selected, list the enzyme name.</td>\n",
       "      <td></td>\n",
       "      <td>specifyOtherCleavageAgent</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>Hispanic or latinoEthnicity</td>\n",
       "      <td>A coded value specifying the self-declared eth...</td>\n",
       "      <td></td>\n",
       "      <td>ethnicGroupCode</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>https://ncithesaurus.nci.nih.gov/ncitbrowser/p...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>TrueDiagnosisStatus</td>\n",
       "      <td>Indicate the disease or condition.</td>\n",
       "      <td>HPO, MONDO, MAXO codes or labels (not listed f...</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/mondo,htt...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>CaptiveCaptivityStatus</td>\n",
       "      <td>The duration of captivity in months for the in...</td>\n",
       "      <td>Unknown,Not collected,Not applicable</td>\n",
       "      <td>captivityDuration</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Individual nonHuman</td>\n",
       "      <td></td>\n",
       "      <td>https://www.animalaudiograms.org/audiogram_met...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117 rows  10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Attribute  \\\n",
       "2                   OtherSampleType   \n",
       "3         OtherMeasurementTechnique   \n",
       "5    OtherTechnologyPlatformVersion   \n",
       "7             OtherPlatformLocation   \n",
       "13            OtherDnaBatchSizeUnit   \n",
       "..                              ...   \n",
       "294     OtherModificationParameters   \n",
       "296             OtherCleavageAgents   \n",
       "345     Hispanic or latinoEthnicity   \n",
       "348             TrueDiagnosisStatus   \n",
       "360          CaptiveCaptivityStatus   \n",
       "\n",
       "                                           Description  \\\n",
       "2       If \"other\" is selected list the type of sample   \n",
       "3    If \"other\" is selected list the name of the me...   \n",
       "5     If \"other\" list the name of the platform version   \n",
       "7    If \"other\" list the name of the platform location   \n",
       "13                     If \"other\" list unit of measure   \n",
       "..                                                 ...   \n",
       "294  If \"other\" is selected, list the custom modifi...   \n",
       "296     If \"other\" is selected, list the enzyme name.    \n",
       "345  A coded value specifying the self-declared eth...   \n",
       "348                 Indicate the disease or condition.   \n",
       "360  The duration of captivity in months for the in...   \n",
       "\n",
       "                                          Valid Values  \\\n",
       "2                                                        \n",
       "3                                                        \n",
       "5                                                        \n",
       "7                                                        \n",
       "13                                                       \n",
       "..                                                 ...   \n",
       "294                                                      \n",
       "296                                                      \n",
       "345                                                      \n",
       "348  HPO, MONDO, MAXO codes or labels (not listed f...   \n",
       "360               Unknown,Not collected,Not applicable   \n",
       "\n",
       "                              DependsOn Properties Required  \\\n",
       "2                     specifySampleType               False   \n",
       "3           specifyMeasurementTechnique               False   \n",
       "5                specifyPlatformVersion               False   \n",
       "7               specifyPlatformLocation               False   \n",
       "13              specifyDNABatchSizeUnit               False   \n",
       "..                                  ...        ...      ...   \n",
       "294  specifyOtherModificationParameters               False   \n",
       "296           specifyOtherCleavageAgent               False   \n",
       "345                     ethnicGroupCode               False   \n",
       "348                           diagnosis               False   \n",
       "360                   captivityDuration               False   \n",
       "\n",
       "                                               Parent DependsOn Component  \\\n",
       "2    bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "3    bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "5    bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "7    bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "13   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "..                                                ...                 ...   \n",
       "294                                        proteomics                       \n",
       "296                                        proteomics                       \n",
       "345                                  Individual Human                       \n",
       "348                                  Individual Human                       \n",
       "360                               Individual nonHuman                       \n",
       "\n",
       "                                                Source Validation Rules  \n",
       "2                                     Sage Bionetworks                   \n",
       "3                                     Sage Bionetworks                   \n",
       "5                                     Sage Bionetworks                   \n",
       "7                                     Sage Bionetworks                   \n",
       "13                                    Sage Bionetworks                   \n",
       "..                                                 ...              ...  \n",
       "294                                   Sage Bionetworks                   \n",
       "296                                   Sage Bionetworks                   \n",
       "345  https://ncithesaurus.nci.nih.gov/ncitbrowser/p...                   \n",
       "348  https://www.ebi.ac.uk/ols/ontologies/mondo,htt...                   \n",
       "360  https://www.animalaudiograms.org/audiogram_met...                   \n",
       "\n",
       "[117 rows x 10 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "equals_df = equals_df.rename(\n",
    "    {\"Attribute\": \"DependsOn\", \"equals_attribute\": \"Attribute\"}, axis=1\n",
    ")[base_cols]\n",
    "\n",
    "equals_df[\"DependsOn Component\"] = \"\"\n",
    "\n",
    "equals_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Update base attribute equals values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "others[\"base_attribute\"] = others[\"equals_series\"].str.split(\"=\", expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "others[\"Properties\"] = \"dataProperty\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dm[dm['Parent'].str.contains('bsSeq ', regex=True)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>base_attribute</th>\n",
       "      <th>value_to_replace</th>\n",
       "      <th>new_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sampleType</td>\n",
       "      <td>Other</td>\n",
       "      <td>OtherSampleType</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>Other</td>\n",
       "      <td>OtherMeasurementTechnique</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>technologyPlatformVersion</td>\n",
       "      <td>Other</td>\n",
       "      <td>OtherTechnologyPlatformVersion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>platformLocation</td>\n",
       "      <td>Other</td>\n",
       "      <td>OtherPlatformLocation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dnaBatchSizeUnit</td>\n",
       "      <td>Other</td>\n",
       "      <td>OtherDnaBatchSizeUnit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>modificationParameters</td>\n",
       "      <td>Other</td>\n",
       "      <td>OtherModificationParameters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>cleavageAgents</td>\n",
       "      <td>Other</td>\n",
       "      <td>OtherCleavageAgents</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>ethnicity</td>\n",
       "      <td>Hispanic or latino</td>\n",
       "      <td>Hispanic or latinoEthnicity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>diagnosisStatus</td>\n",
       "      <td>True</td>\n",
       "      <td>TrueDiagnosisStatus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>captivityStatus</td>\n",
       "      <td>Captive</td>\n",
       "      <td>CaptiveCaptivityStatus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                base_attribute    value_to_replace  \\\n",
       "0                   sampleType               Other   \n",
       "1         measurementTechnique               Other   \n",
       "2    technologyPlatformVersion               Other   \n",
       "3             platformLocation               Other   \n",
       "4             dnaBatchSizeUnit               Other   \n",
       "..                         ...                 ...   \n",
       "112     modificationParameters               Other   \n",
       "113             cleavageAgents               Other   \n",
       "114                  ethnicity  Hispanic or latino   \n",
       "115            diagnosisStatus                True   \n",
       "116            captivityStatus             Captive   \n",
       "\n",
       "                          new_value  \n",
       "0                   OtherSampleType  \n",
       "1         OtherMeasurementTechnique  \n",
       "2    OtherTechnologyPlatformVersion  \n",
       "3             OtherPlatformLocation  \n",
       "4             OtherDnaBatchSizeUnit  \n",
       "..                              ...  \n",
       "112     OtherModificationParameters  \n",
       "113             OtherCleavageAgents  \n",
       "114     Hispanic or latinoEthnicity  \n",
       "115             TrueDiagnosisStatus  \n",
       "116          CaptiveCaptivityStatus  \n",
       "\n",
       "[117 rows x 3 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = others[\"equals_series\"].str.split(\"=\", expand=True)\n",
    "temp = temp.apply(lambda x: x.str.strip(), axis=1)\n",
    "temp = temp.rename({0: \"base_attribute\", 1: \"value_to_replace\"}, axis=1)\n",
    "temp[\"value_to_replace\"] = temp[\"value_to_replace\"].str.capitalize()\n",
    "\n",
    "# new value\n",
    "temp[\"new_value\"] = others[\"equals_attribute\"]\n",
    "\n",
    "temp.reset_index(drop=True, inplace=True)\n",
    "\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = temp.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>base_attribute</th>\n",
       "      <th>value_to_replace</th>\n",
       "      <th>new_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [base_attribute, value_to_replace, new_value]\n",
       "Index: []"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp[temp[\"new_value\"] == \"YesHasIonizationSource?\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sampleType': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherSampleType'},\n",
       " 'measurementTechnique': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherMeasurementTechnique'},\n",
       " 'technologyPlatformVersion': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherTechnologyPlatformVersion'},\n",
       " 'platformLocation': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherPlatformLocation'},\n",
       " 'dnaBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherDnaBatchSizeUnit'},\n",
       " 'libraryPrep': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherLibraryPrep'},\n",
       " 'libraryPreparationMethod': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherLibraryPreparationMethod'},\n",
       " 'libraryVersion': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherLibraryVersion'},\n",
       " 'sequencingBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherSequencingBatchSizeUnit'},\n",
       " 'readLengthUnits': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherReadLengthUnits'},\n",
       " 'repositoryName': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherRepositoryName'},\n",
       " 'transcriptType': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherTranscriptType'},\n",
       " 'rnaBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherRnaBatchSizeUnit'},\n",
       " 'useReagent': {'pattern': re.compile(r'(Yes)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'YesUseReagent'},\n",
       " 'useTreatment': {'pattern': re.compile(r'(Yes)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'YesUseTreatment'},\n",
       " 'Metabolomics HumanmeasurementTechnique': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherMetabolomics HumanmeasurementTechnique'},\n",
       " 'msAnalyzerTypeMS1': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherMsAnalyzerTypeMS1'},\n",
       " 'msAnalyzerTypeMS2': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherMsAnalyzerTypeMS2'},\n",
       " 'msAssayTechnique': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherMsAssayTechnique'},\n",
       " 'msTarget': {'pattern': re.compile(r'(Targeted)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TargetedMsTarget'},\n",
       " 'msAnalyteType': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherMsAnalyteType'},\n",
       " 'msInstrumentModel': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherMsInstrumentModel'},\n",
       " 'sampleIntdocution': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherSampleIntdocution'},\n",
       " 'databaseName': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherDatabaseName'},\n",
       " 'databaseSource': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherDatabaseSource'},\n",
       " 'hasIonizationSource': {'pattern': re.compile(r'(Yes)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'YesHasIonizationSource'},\n",
       " 'vaccumPressureUnits': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherVaccumPressureUnits'},\n",
       " 'lensVoltageUnits': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherLensVoltageUnits'},\n",
       " 'hasAssayControl': {'pattern': re.compile(r'(Yes)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'YesHasAssayControl'},\n",
       " 'controlType': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherControlType'},\n",
       " 'experimentalBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherExperimentalBatchSizeUnit'},\n",
       " 'batchSizeUnit': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherBatchSizeUnit'},\n",
       " 'sampleBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherSampleBatchSizeUnit'},\n",
       " 'acquisitionBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherAcquisitionBatchSizeUnit'},\n",
       " 'ProcessingBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherProcessingBatchSizeUnit'},\n",
       " 'extractionMethod': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherExtractionMethod'},\n",
       " 'proteomicsAssayType': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherProteomicsAssayType'},\n",
       " 'digestionMethod': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherDigestionMethod'},\n",
       " 'modificationParameters': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherModificationParameters'},\n",
       " 'cleavageAgents': {'pattern': re.compile(r'(Other)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'OtherCleavageAgents'},\n",
       " 'ethnicity': {'pattern': re.compile(r'(Hispanic or latino)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'Hispanic or latinoEthnicity'},\n",
       " 'diagnosisStatus': {'pattern': re.compile(r'(True)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TrueDiagnosisStatus'},\n",
       " 'captivityStatus': {'pattern': re.compile(r'(Captive)',\n",
       "  re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'CaptiveCaptivityStatus'}}"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "replacements = {}\n",
    "\n",
    "for i, x in temp.iterrows():\n",
    "    ba, vtr, nv = x\n",
    "    replacements[ba] = {\n",
    "        \"pattern\": re.compile(\"(\" + vtr + \")\", flags=re.IGNORECASE),\n",
    "        \"repl\": nv,\n",
    "    }\n",
    "\n",
    "replacements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "attribute:  sampleType\n",
      "Index:  [1, 47, 73, 110, 141, 162, 228, 260]\n",
      "--------------------\n",
      "attribute:  measurementTechnique\n",
      "measurementTechnique\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  technologyPlatformVersion\n",
      "Index:  [4, 50, 76, 113, 165, 231, 263]\n",
      "--------------------\n",
      "attribute:  platformLocation\n",
      "Index:  [6, 52, 78, 115, 167, 233, 265]\n",
      "--------------------\n",
      "attribute:  dnaBatchSizeUnit\n",
      "Index:  [12, 244]\n",
      "--------------------\n",
      "attribute:  libraryPrep\n",
      "Index:  [15, 96, 127, 246]\n",
      "--------------------\n",
      "attribute:  libraryPreparationMethod\n",
      "Index:  [17, 25, 61, 98, 129, 248]\n",
      "--------------------\n",
      "attribute:  libraryVersion\n",
      "Index:  [19, 27, 63, 100, 131, 250]\n",
      "--------------------\n",
      "attribute:  sequencingBatchSizeUnit\n",
      "Index:  [23, 94, 240]\n",
      "--------------------\n",
      "attribute:  readLengthUnits\n",
      "Index:  [32, 68, 105, 136, 255]\n",
      "--------------------\n",
      "attribute:  repositoryName\n",
      "Index:  [55, 81, 118]\n",
      "--------------------\n",
      "attribute:  transcriptType\n",
      "Index:  [57, 83, 120]\n",
      "--------------------\n",
      "attribute:  rnaBatchSizeUnit\n",
      "Index:  [89]\n",
      "--------------------\n",
      "attribute:  useReagent\n",
      "Index:  [143]\n",
      "--------------------\n",
      "attribute:  useTreatment\n",
      "Index:  [151]\n",
      "--------------------\n",
      "attribute:  Metabolomics HumanmeasurementTechnique\n",
      "Metabolomics HumanmeasurementTechnique\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  msAnalyzerTypeMS1\n",
      "Index:  [170, 268]\n",
      "--------------------\n",
      "attribute:  msAnalyzerTypeMS2\n",
      "Index:  [172, 270]\n",
      "--------------------\n",
      "attribute:  msAssayTechnique\n",
      "Index:  [174, 272]\n",
      "--------------------\n",
      "attribute:  msTarget\n",
      "Index:  [169, 267]\n",
      "--------------------\n",
      "attribute:  msAnalyteType\n",
      "Index:  [176, 274]\n",
      "--------------------\n",
      "attribute:  msInstrumentModel\n",
      "Index:  [178, 278]\n",
      "--------------------\n",
      "attribute:  sampleIntdocution\n",
      "sampleIntdocution\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  databaseName\n",
      "Index:  [187, 287]\n",
      "--------------------\n",
      "attribute:  databaseSource\n",
      "Index:  [189, 289]\n",
      "--------------------\n",
      "attribute:  hasIonizationSource\n",
      "Index:  [193]\n",
      "--------------------\n",
      "attribute:  vaccumPressureUnits\n",
      "vaccumPressureUnits\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  lensVoltageUnits\n",
      "lensVoltageUnits\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  hasAssayControl\n",
      "Index:  [203]\n",
      "--------------------\n",
      "attribute:  controlType\n",
      "Index:  [204]\n",
      "--------------------\n",
      "attribute:  experimentalBatchSizeUnit\n",
      "experimentalBatchSizeUnit\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  batchSizeUnit\n",
      "Index:  [210]\n",
      "--------------------\n",
      "attribute:  sampleBatchSizeUnit\n",
      "Index:  [216]\n",
      "--------------------\n",
      "attribute:  acquisitionBatchSizeUnit\n",
      "Index:  [220]\n",
      "--------------------\n",
      "attribute:  ProcessingBatchSizeUnit\n",
      "Index:  [224]\n",
      "--------------------\n",
      "attribute:  extractionMethod\n",
      "Index:  [182, 235]\n",
      "--------------------\n",
      "attribute:  proteomicsAssayType\n",
      "proteomicsAssayType\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  digestionMethod\n",
      "Index:  [280]\n",
      "--------------------\n",
      "attribute:  modificationParameters\n",
      "Index:  [293]\n",
      "--------------------\n",
      "attribute:  cleavageAgents\n",
      "Index:  [295]\n",
      "--------------------\n",
      "attribute:  ethnicity\n",
      "Index:  [344]\n",
      "--------------------\n",
      "attribute:  diagnosisStatus\n",
      "Index:  [347]\n",
      "--------------------\n",
      "attribute:  captivityStatus\n",
      "Index:  [359]\n",
      "--------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\nlee\\anaconda3\\envs\\schematic\\lib\\site-packages\\ipykernel_launcher.py:4: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.\n",
      "  after removing the cwd from sys.path.\n"
     ]
    }
   ],
   "source": [
    "for attribute in replacements.keys():\n",
    "    dm = code_equals_values(dm, replacements, attribute)\n",
    "    print(\"-\" * 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with pd.option_context('display.max_rows', None, 'display.max_colwidth', 0):\n",
    "#     display(\n",
    "#         dm[dm['Attribute'].str.contains(\n",
    "#             'proteomicsAssayType', flags=re.IGNORECASE)]\n",
    "#     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "equals_df[\"Properties\"] = \"validValue\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add new attributes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm[\"Properties\"] = \"dataProperty\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm.update(others[base_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm = pd.concat([dm, equals_df], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Do not need\n",
    "dm[\"DependsOn Component\"] = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean Up Valid Values and Add New Attributes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\nlee\\anaconda3\\envs\\schematic\\lib\\site-packages\\ipykernel_launcher.py:3: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    }
   ],
   "source": [
    "# valid values that contain other\n",
    "pure_others = dm[dm[\"Valid Values\"].str.contains(\"([Oo]ther$)\", flags=re.IGNORECASE)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\nlee\\anaconda3\\envs\\schematic\\lib\\site-packages\\ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "pure_others[\"replacement_value\"] = pure_others[\"Attribute\"].apply(\n",
    "    lambda x: \"Other\" + (x[0].upper() + x[1:])\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "regex_dict = {}\n",
    "\n",
    "pattern = \"([Oo]ther$)\"\n",
    "\n",
    "for i, r in pure_others.iterrows():\n",
    "    regex_dict[r[\"Attribute\"]] = {\n",
    "        \"pattern\": re.compile(pattern, flags=re.IGNORECASE),\n",
    "        \"repl\": r[\"replacement_value\"],\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"conversionRatioUnits\": {\n",
      "        \"pattern\": \"re.compile('([Oo]ther$)', re.IGNORECASE)\",\n",
      "        \"repl\": \"OtherConversionRatioUnits\"\n",
      "    },\n",
      "    \"vacuumPressureUnit\": {\n",
      "        \"pattern\": \"re.compile('([Oo]ther$)', re.IGNORECASE)\",\n",
      "        \"repl\": \"OtherVacuumPressureUnit\"\n",
      "    },\n",
      "    \"lensVoltagesUnit \": {\n",
      "        \"pattern\": \"re.compile('([Oo]ther$)', re.IGNORECASE)\",\n",
      "        \"repl\": \"OtherLensVoltagesUnit \"\n",
      "    },\n",
      "    \"experiementalBatchSizeUnit\": {\n",
      "        \"pattern\": \"re.compile('([Oo]ther$)', re.IGNORECASE)\",\n",
      "        \"repl\": \"OtherExperiementalBatchSizeUnit\"\n",
      "    },\n",
      "    \"YesHasIonizationSource\": {\n",
      "        \"pattern\": \"re.compile('([Oo]ther$)', re.IGNORECASE)\",\n",
      "        \"repl\": \"OtherYesHasIonizationSource\"\n",
      "    }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "temp_dict = {}\n",
    "\n",
    "for i, r in pure_others.iterrows():\n",
    "    temp_dict[r[\"Attribute\"]] = {\n",
    "        \"pattern\": str(re.compile(pattern, flags=re.IGNORECASE)),\n",
    "        \"repl\": r[\"replacement_value\"],\n",
    "    }\n",
    "\n",
    "json_formatted = json.dumps(temp_dict, indent=4)\n",
    "print(json_formatted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "attribute:  conversionRatioUnits\n",
      "Index:  [42]\n",
      "--------------------\n",
      "attribute:  vacuumPressureUnit\n",
      "Index:  [196]\n",
      "--------------------\n",
      "attribute:  lensVoltagesUnit \n",
      "Index:  [199]\n",
      "--------------------\n",
      "attribute:  experiementalBatchSizeUnit\n",
      "Index:  [207]\n",
      "--------------------\n",
      "attribute:  YesHasIonizationSource\n",
      "Index:  [432, 433, 434, 436, 437, 439, 440]\n",
      "--------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\nlee\\anaconda3\\envs\\schematic\\lib\\site-packages\\ipykernel_launcher.py:4: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.\n",
      "  after removing the cwd from sys.path.\n"
     ]
    }
   ],
   "source": [
    "for attribute in regex_dict.keys():\n",
    "    dm = code_equals_values(dm, regex_dict, attribute)\n",
    "    print(\"-\" * 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm[\"Valid Values\"] = dm[\"Valid Values\"].str.replace(\n",
    "    re.escape(\"(falseFalseFALSEtrueTrueTRUE)\"), \"\", regex=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with pd.option_context('display.max_rows', None, 'display.max_colwidth', 0):\n",
    "#     display(dm[dm['Valid Values'].str.contains(\n",
    "#         re.escape('(not listed for purposes of this RFC)'))].sort_values(by='Source'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_list_from_lists(main_list, new_list):\n",
    "    try:\n",
    "        main_list = main_list + list\n",
    "    except:\n",
    "        pass\n",
    "    return main_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "split_valid_values = dm[\"Valid Values\"].str.split(\",\")\n",
    "new_valid_values = [i for i in split_valid_values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = []\n",
    "for nv in new_valid_values:\n",
    "    # print(nv)\n",
    "    try:\n",
    "        test = test + nv\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "test = list(np.unique(test))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_values = \",\".join(dm[\"Valid Values\"])\n",
    "valid_values = valid_values.split(\",\")\n",
    "valid_values = list(np.unique(valid_values))\n",
    "valid_values = [v.strip() for v in valid_values if len(v) > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'FALSE': {'False': 100},\n",
       " 'False': {'FALSE': 100},\n",
       " 'Lipid': {'lipid': 100},\n",
       " 'Not Specified': {'Not specified': 100},\n",
       " 'Not specified': {'Not Specified': 100},\n",
       " 'Plasma': {'plasma': 100},\n",
       " 'Protein': {'protein': 100},\n",
       " 'Saliva': {'saliva': 100},\n",
       " 'Serum': {'serum': 100},\n",
       " 'Sputum': {'sputum': 100},\n",
       " 'TRUE': {'True': 100},\n",
       " 'True': {'TRUE': 100},\n",
       " 'Urine': {'urine': 100},\n",
       " 'lipid': {'Lipid': 100},\n",
       " 'plasma': {'Plasma': 100},\n",
       " 'protein': {'Protein': 100},\n",
       " 'saliva': {'Saliva': 100},\n",
       " 'serum': {'Serum': 100},\n",
       " 'sputum': {'Sputum': 100},\n",
       " 'urine': {'Urine': 100}}"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fuzzy matching to find misspellings\n",
    "# Fuzzy matching\n",
    "from thefuzz import fuzz\n",
    "\n",
    "scores = {}\n",
    "for v in valid_values:\n",
    "    scores[v] = {}\n",
    "    for v2 in valid_values:\n",
    "        if v == v2:\n",
    "            next\n",
    "        else:\n",
    "            score = fuzz.ratio(v.lower(), v2.lower())\n",
    "            if score == 100:\n",
    "                scores[v][v2] = score\n",
    "    if len(scores[v]) == 0:\n",
    "        scores.pop(v)\n",
    "\n",
    "scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_values_recoded = []\n",
    "for v in scores.values():\n",
    "    new_values_recoded.append(list(v.keys())[0].lower())\n",
    "\n",
    "new_values_recoded = np.unique(new_values_recoded)\n",
    "\n",
    "recoder_valid_values = []\n",
    "for nv in new_values_recoded:\n",
    "    value_add = {re.compile(nv, flags=re.IGNORECASE): nv}\n",
    "    recoder_valid_values.append(value_add)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{re.compile(r'false', re.IGNORECASE|re.UNICODE): 'false'},\n",
       " {re.compile(r'lipid', re.IGNORECASE|re.UNICODE): 'lipid'},\n",
       " {re.compile(r'not specified', re.IGNORECASE|re.UNICODE): 'not specified'},\n",
       " {re.compile(r'plasma', re.IGNORECASE|re.UNICODE): 'plasma'},\n",
       " {re.compile(r'protein', re.IGNORECASE|re.UNICODE): 'protein'},\n",
       " {re.compile(r'saliva', re.IGNORECASE|re.UNICODE): 'saliva'},\n",
       " {re.compile(r'serum', re.IGNORECASE|re.UNICODE): 'serum'},\n",
       " {re.compile(r'sputum', re.IGNORECASE|re.UNICODE): 'sputum'},\n",
       " {re.compile(r'true', re.IGNORECASE|re.UNICODE): 'true'},\n",
       " {re.compile(r'urine', re.IGNORECASE|re.UNICODE): 'urine'}]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recoder_valid_values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cleanup valid values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_list(x):\n",
    "    try:\n",
    "        result = \",\".join([y.strip() for y in x])\n",
    "        return result\n",
    "    except:\n",
    "        return \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "recoder_valid_values = {\n",
    "    re.compile(\"Not Specified\", flags=re.IGNORECASE): \"Not Specified\",\n",
    "    re.compile(\"(Other$)\", flags=re.IGNORECASE): \"Other\",\n",
    "    re.compile(\"lipid\", re.IGNORECASE): \"Lipid\",\n",
    "    re.compile(\"plasma\", re.IGNORECASE): \"Plasma\",\n",
    "    re.compile(\"protein\", re.IGNORECASE): \"Protein\",\n",
    "    re.compile(\"saliva\", re.IGNORECASE): \"Saliva\",\n",
    "    re.compile(\"serum\", re.IGNORECASE): \"Serum\",\n",
    "    re.compile(\"sputum\", re.IGNORECASE): \"Sputum\",\n",
    "    re.compile(\"urine\", re.IGNORECASE): \"Urine\",\n",
    "    re.compile(\n",
    "        \"(^0x Visium Spatial Gene Expression)\"\n",
    "    ): \"10x Visium Spatial Gene Expression\",\n",
    "    re.compile(\"falseFalseFALSEtrueTrueTRUE\t\"): \"TRUE, FALSE\",\n",
    "    re.compile(\"TRUE|TRUEDiagnosisStatus\", re.IGNORECASE): \"TRUE\",\n",
    "    re.compile(\"TRUEDiagnosisStatus\", re.IGNORECASE): \"TRUE\",\n",
    "    re.compile(\"FALSE\", re.IGNORECASE): \"FALSE\",\n",
    "    re.compile(\"$f^\", re.IGNORECASE): \"F\",\n",
    "    re.compile(\"UnknownNot collected\"): \"Unknown, Not collected\",\n",
    "    re.compile(r\"\\u200b\\u200b\"): \"\",\n",
    "    re.compile(\n",
    "        \"The Health,Aging,and Body Composition Study \\(HealthABC\\)\"\n",
    "    ): \"The Health and Aging and Body Composition Study (HealthABC)\",\n",
    "    re.compile(\"Not Hispanic or latinoEthnicity\"): \"Not Hispanic or latino\",\n",
    "    re.compile(\"Hispanic or latinoEthnicity\"): \"Hispanic or latino\",\n",
    "    re.compile(\n",
    "        \"HPO, MONDO, MAXO codes or labels \\(not listed for purposes of this RFC\\)\"\n",
    "    ): \"HPO and MONDO and MAXO codes or labels (not listed for purposes of this RFC)\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aging</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Baker pool</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cardiovascular Health Study (CHS)</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Denmark Family</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FALSE</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>ul</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>440</th>\n",
       "      <td>umol/l</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>441</th>\n",
       "      <td>units/ml</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442</th>\n",
       "      <td>urine</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>Affymetrix Genome-Wide Human SNP 5.0 Array</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>414 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Attribute  Properties Required\n",
       "0                                           Aging  validValue    False\n",
       "1                                      Baker pool  validValue    False\n",
       "2               Cardiovascular Health Study (CHS)  validValue    False\n",
       "3                                  Denmark Family  validValue    False\n",
       "4                                           FALSE  validValue    False\n",
       "..                                            ...         ...      ...\n",
       "439                                            ul  validValue    False\n",
       "440                                        umol/l  validValue    False\n",
       "441                                      units/ml  validValue    False\n",
       "442                                         urine  validValue    False\n",
       "443  Affymetrix Genome-Wide Human SNP 5.0 Array  validValue    False\n",
       "\n",
       "[414 rows x 3 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm[\"Valid Values\"] = (\n",
    "    dm[\"Valid Values\"]\n",
    "    .replace(recoder_valid_values, regex=True)\n",
    "    .str.split(\",\")\n",
    "    .apply(lambda x: clean_list(x))\n",
    ")\n",
    "\n",
    "valid_values = list(np.unique(\",\".join(test).split(\",\")))\n",
    "\n",
    "valid_values = [v.strip() for v in valid_values if len(v) > 0]\n",
    "\n",
    "valid_values_df = pd.DataFrame(\n",
    "    {\n",
    "        \"Attribute\": pd.Series(valid_values),\n",
    "        \"Properties\": pd.Series([\"validValue\" for v in valid_values]),\n",
    "        \"Required\": pd.Series([\"False\" for v in valid_values]),\n",
    "    }\n",
    ")\n",
    "\n",
    "valid_values_df = valid_values_df[\n",
    "    ~valid_values_df[\"Attribute\"].isin(dm[\"Attribute\"].tolist())\n",
    "]\n",
    "\n",
    "valid_values_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "# adding valid values found in attribute columns\n",
    "dm2 = pd.concat([dm, valid_values_df], axis=0, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'f': {'F': 100}, 'F': {'f': 100}}"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fuzzy matching to find misspellings\n",
    "# Fuzzy matching\n",
    "\n",
    "from thefuzz import fuzz\n",
    "\n",
    "valid_values = dm2[\"Attribute\"].replace(recoder_valid_values, regex=True).tolist()\n",
    "\n",
    "scores = {}\n",
    "for v in valid_values:\n",
    "    scores[v] = {}\n",
    "    for v2 in valid_values:\n",
    "        if v == v2:\n",
    "            next\n",
    "        else:\n",
    "            score = fuzz.ratio(v.lower(), v2.lower())\n",
    "            if score == 100:\n",
    "                scores[v][v2] = score\n",
    "    if len(scores[v]) == 0:\n",
    "        scores.pop(v)\n",
    "\n",
    "scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm2[[\"Attribute\", \"DependsOn\"]] = dm2[[\"Attribute\", \"DependsOn\"]].apply(\n",
    "    lambda x: x.replace(recoder_valid_values, regex=True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>OtherLibraryPreparationMethod</td>\n",
       "      <td>If \"other\" list the name of the library prepar...</td>\n",
       "      <td></td>\n",
       "      <td>specifyLibraryPreparationMethod</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>OtherLibraryPreparationMethod</td>\n",
       "      <td>If \"other\" list the name of the library prepar...</td>\n",
       "      <td></td>\n",
       "      <td>specifyLibraryPreparationMethod</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>OtherLibraryVersion</td>\n",
       "      <td>If \"other\" list the name of the library version</td>\n",
       "      <td></td>\n",
       "      <td>specifyLibraryVersion</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>OtherLibraryVersion</td>\n",
       "      <td>If \"other\" list the name of the library version</td>\n",
       "      <td></td>\n",
       "      <td>specifyLibraryVersion</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>dnaBatchID</td>\n",
       "      <td>DNA isolation batch</td>\n",
       "      <td>Other,Unknown,Not collected,Not applicable</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>dnaBatchID</td>\n",
       "      <td>DNA isolation batch identifier, provided by th...</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not Speci...</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>libraryPreparationMethod</td>\n",
       "      <td>Method by which library was prepared</td>\n",
       "      <td>10x,Accel-NGS 2S Plus,Accel-NGS Methyl-Seq,CEL...</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>libraryPreparationMethod</td>\n",
       "      <td>Method by which library was prepared</td>\n",
       "      <td>10x,Accel-NGS 2S Plus,Accel-NGS Methyl-Seq,CEL...</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>libraryVersion</td>\n",
       "      <td>Library Version: for example, rnaSeq 10x libra...</td>\n",
       "      <td>OtherLibraryVersion,Unknown,Not collected,Not ...</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>libraryVersion</td>\n",
       "      <td>Library Version: for example, rnaSeq 10x libra...</td>\n",
       "      <td>OtherLibraryVersion,Unknown,Not collected,Not ...</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>specifyLibraryPreparationMethod</td>\n",
       "      <td>If \"other\" list the name of the library prepar...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>specifyLibraryPreparationMethod</td>\n",
       "      <td>If \"other\" list the name of the library prepar...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>specifyLibraryVersion</td>\n",
       "      <td>If \"other\" list the name of the library version</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>specifyLibraryVersion</td>\n",
       "      <td>If \"other\" list the name of the library version</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>10x Visium Spatial Gene Expression</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>10x Visium Spatial Gene Expression</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482</th>\n",
       "      <td>FALSE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>FALSE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>FALSE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639</th>\n",
       "      <td>Lipid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>825</th>\n",
       "      <td>Lipid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>Not Specified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>690</th>\n",
       "      <td>Not Specified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>Plasma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>857</th>\n",
       "      <td>Plasma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>Prefer not to answer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731</th>\n",
       "      <td>Prefer not to answer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>734</th>\n",
       "      <td>Protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>Protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749</th>\n",
       "      <td>Saliva</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864</th>\n",
       "      <td>Saliva</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>752</th>\n",
       "      <td>Serum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866</th>\n",
       "      <td>Serum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>760</th>\n",
       "      <td>Sputum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>873</th>\n",
       "      <td>Sputum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771</th>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778</th>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>491</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>791</th>\n",
       "      <td>Urine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>890</th>\n",
       "      <td>Urine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>pairedEnd</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>850</th>\n",
       "      <td>pairedEnd</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Attribute  \\\n",
       "367       OtherLibraryPreparationMethod   \n",
       "370       OtherLibraryPreparationMethod   \n",
       "368                 OtherLibraryVersion   \n",
       "371                 OtherLibraryVersion   \n",
       "9                            dnaBatchID   \n",
       "10                           dnaBatchID   \n",
       "17             libraryPreparationMethod   \n",
       "25             libraryPreparationMethod   \n",
       "19                       libraryVersion   \n",
       "27                       libraryVersion   \n",
       "18      specifyLibraryPreparationMethod   \n",
       "26      specifyLibraryPreparationMethod   \n",
       "20                specifyLibraryVersion   \n",
       "28                specifyLibraryVersion   \n",
       "495  10x Visium Spatial Gene Expression   \n",
       "497  10x Visium Spatial Gene Expression   \n",
       "482                               FALSE   \n",
       "483                               FALSE   \n",
       "568                               FALSE   \n",
       "639                               Lipid   \n",
       "825                               Lipid   \n",
       "687                       Not Specified   \n",
       "690                       Not Specified   \n",
       "721                              Plasma   \n",
       "857                              Plasma   \n",
       "486                Prefer not to answer   \n",
       "731                Prefer not to answer   \n",
       "734                             Protein   \n",
       "859                             Protein   \n",
       "749                              Saliva   \n",
       "864                              Saliva   \n",
       "752                               Serum   \n",
       "866                               Serum   \n",
       "760                              Sputum   \n",
       "873                              Sputum   \n",
       "771                                TRUE   \n",
       "778                                TRUE   \n",
       "491                             Unknown   \n",
       "788                             Unknown   \n",
       "791                               Urine   \n",
       "890                               Urine   \n",
       "493                           pairedEnd   \n",
       "850                           pairedEnd   \n",
       "\n",
       "                                           Description  \\\n",
       "367  If \"other\" list the name of the library prepar...   \n",
       "370  If \"other\" list the name of the library prepar...   \n",
       "368    If \"other\" list the name of the library version   \n",
       "371    If \"other\" list the name of the library version   \n",
       "9                                  DNA isolation batch   \n",
       "10   DNA isolation batch identifier, provided by th...   \n",
       "17                Method by which library was prepared   \n",
       "25                Method by which library was prepared   \n",
       "19   Library Version: for example, rnaSeq 10x libra...   \n",
       "27   Library Version: for example, rnaSeq 10x libra...   \n",
       "18   If \"other\" list the name of the library prepar...   \n",
       "26   If \"other\" list the name of the library prepar...   \n",
       "20     If \"other\" list the name of the library version   \n",
       "28     If \"other\" list the name of the library version   \n",
       "495                                                NaN   \n",
       "497                                                NaN   \n",
       "482                                                NaN   \n",
       "483                                                NaN   \n",
       "568                                                NaN   \n",
       "639                                                NaN   \n",
       "825                                                NaN   \n",
       "687                                                NaN   \n",
       "690                                                NaN   \n",
       "721                                                NaN   \n",
       "857                                                NaN   \n",
       "486                                                NaN   \n",
       "731                                                NaN   \n",
       "734                                                NaN   \n",
       "859                                                NaN   \n",
       "749                                                NaN   \n",
       "864                                                NaN   \n",
       "752                                                NaN   \n",
       "866                                                NaN   \n",
       "760                                                NaN   \n",
       "873                                                NaN   \n",
       "771                                                NaN   \n",
       "778                                                NaN   \n",
       "491                                                NaN   \n",
       "788                                                NaN   \n",
       "791                                                NaN   \n",
       "890                                                NaN   \n",
       "493                                                NaN   \n",
       "850                                                NaN   \n",
       "\n",
       "                                          Valid Values  \\\n",
       "367                                                      \n",
       "370                                                      \n",
       "368                                                      \n",
       "371                                                      \n",
       "9           Other,Unknown,Not collected,Not applicable   \n",
       "10   Unknown,Not collected,Not applicable,Not Speci...   \n",
       "17   10x,Accel-NGS 2S Plus,Accel-NGS Methyl-Seq,CEL...   \n",
       "25   10x,Accel-NGS 2S Plus,Accel-NGS Methyl-Seq,CEL...   \n",
       "19   OtherLibraryVersion,Unknown,Not collected,Not ...   \n",
       "27   OtherLibraryVersion,Unknown,Not collected,Not ...   \n",
       "18                                                       \n",
       "26                                                       \n",
       "20                                                       \n",
       "28                                                       \n",
       "495                                                NaN   \n",
       "497                                                NaN   \n",
       "482                                                NaN   \n",
       "483                                                NaN   \n",
       "568                                                NaN   \n",
       "639                                                NaN   \n",
       "825                                                NaN   \n",
       "687                                                NaN   \n",
       "690                                                NaN   \n",
       "721                                                NaN   \n",
       "857                                                NaN   \n",
       "486                                                NaN   \n",
       "731                                                NaN   \n",
       "734                                                NaN   \n",
       "859                                                NaN   \n",
       "749                                                NaN   \n",
       "864                                                NaN   \n",
       "752                                                NaN   \n",
       "866                                                NaN   \n",
       "760                                                NaN   \n",
       "873                                                NaN   \n",
       "771                                                NaN   \n",
       "778                                                NaN   \n",
       "491                                                NaN   \n",
       "788                                                NaN   \n",
       "791                                                NaN   \n",
       "890                                                NaN   \n",
       "493                                                NaN   \n",
       "850                                                NaN   \n",
       "\n",
       "                           DependsOn    Properties Required  \\\n",
       "367  specifyLibraryPreparationMethod    validValue    False   \n",
       "370  specifyLibraryPreparationMethod    validValue    False   \n",
       "368            specifyLibraryVersion    validValue    False   \n",
       "371            specifyLibraryVersion    validValue    False   \n",
       "9                                     dataProperty     True   \n",
       "10                                    dataProperty     True   \n",
       "17                                    dataProperty     True   \n",
       "25                                    dataProperty     True   \n",
       "19                                    dataProperty     True   \n",
       "27                                    dataProperty     True   \n",
       "18                                    dataProperty    False   \n",
       "26                                    dataProperty    False   \n",
       "20                                    dataProperty    False   \n",
       "28                                    dataProperty    False   \n",
       "495                              NaN    validValue    False   \n",
       "497                              NaN    validValue    False   \n",
       "482                              NaN    validValue    False   \n",
       "483                              NaN    validValue    False   \n",
       "568                              NaN    validValue    False   \n",
       "639                              NaN    validValue    False   \n",
       "825                              NaN    validValue    False   \n",
       "687                              NaN    validValue    False   \n",
       "690                              NaN    validValue    False   \n",
       "721                              NaN    validValue    False   \n",
       "857                              NaN    validValue    False   \n",
       "486                              NaN    validValue    False   \n",
       "731                              NaN    validValue    False   \n",
       "734                              NaN    validValue    False   \n",
       "859                              NaN    validValue    False   \n",
       "749                              NaN    validValue    False   \n",
       "864                              NaN    validValue    False   \n",
       "752                              NaN    validValue    False   \n",
       "866                              NaN    validValue    False   \n",
       "760                              NaN    validValue    False   \n",
       "873                              NaN    validValue    False   \n",
       "771                              NaN    validValue    False   \n",
       "778                              NaN    validValue    False   \n",
       "491                              NaN    validValue    False   \n",
       "788                              NaN    validValue    False   \n",
       "791                              NaN    validValue    False   \n",
       "890                              NaN    validValue    False   \n",
       "493                              NaN    validValue    False   \n",
       "850                              NaN    validValue    False   \n",
       "\n",
       "                                               Parent DependsOn Component  \\\n",
       "367  bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "370  bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "368  bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "371  bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "9    bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "10   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "17   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "25   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "19   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "27   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "18   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "26   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "20   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "28   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "495                                               NaN                 NaN   \n",
       "497                                               NaN                 NaN   \n",
       "482                                               NaN                 NaN   \n",
       "483                                               NaN                 NaN   \n",
       "568                                               NaN                 NaN   \n",
       "639                                               NaN                 NaN   \n",
       "825                                               NaN                 NaN   \n",
       "687                                               NaN                 NaN   \n",
       "690                                               NaN                 NaN   \n",
       "721                                               NaN                 NaN   \n",
       "857                                               NaN                 NaN   \n",
       "486                                               NaN                 NaN   \n",
       "731                                               NaN                 NaN   \n",
       "734                                               NaN                 NaN   \n",
       "859                                               NaN                 NaN   \n",
       "749                                               NaN                 NaN   \n",
       "864                                               NaN                 NaN   \n",
       "752                                               NaN                 NaN   \n",
       "866                                               NaN                 NaN   \n",
       "760                                               NaN                 NaN   \n",
       "873                                               NaN                 NaN   \n",
       "771                                               NaN                 NaN   \n",
       "778                                               NaN                 NaN   \n",
       "491                                               NaN                 NaN   \n",
       "788                                               NaN                 NaN   \n",
       "791                                               NaN                 NaN   \n",
       "890                                               NaN                 NaN   \n",
       "493                                               NaN                 NaN   \n",
       "850                                               NaN                 NaN   \n",
       "\n",
       "               Source Validation Rules  \n",
       "367  Sage Bionetworks                   \n",
       "370  Sage Bionetworks                   \n",
       "368  Sage Bionetworks                   \n",
       "371  Sage Bionetworks                   \n",
       "9    Sage Bionetworks                   \n",
       "10   Sage Bionetworks                   \n",
       "17   Sage Bionetworks                   \n",
       "25   Sage Bionetworks                   \n",
       "19   Sage Bionetworks                   \n",
       "27   Sage Bionetworks                   \n",
       "18   Sage Bionetworks                   \n",
       "26   Sage Bionetworks                   \n",
       "20   Sage Bionetworks                   \n",
       "28   Sage Bionetworks                   \n",
       "495               NaN              NaN  \n",
       "497               NaN              NaN  \n",
       "482               NaN              NaN  \n",
       "483               NaN              NaN  \n",
       "568               NaN              NaN  \n",
       "639               NaN              NaN  \n",
       "825               NaN              NaN  \n",
       "687               NaN              NaN  \n",
       "690               NaN              NaN  \n",
       "721               NaN              NaN  \n",
       "857               NaN              NaN  \n",
       "486               NaN              NaN  \n",
       "731               NaN              NaN  \n",
       "734               NaN              NaN  \n",
       "859               NaN              NaN  \n",
       "749               NaN              NaN  \n",
       "864               NaN              NaN  \n",
       "752               NaN              NaN  \n",
       "866               NaN              NaN  \n",
       "760               NaN              NaN  \n",
       "873               NaN              NaN  \n",
       "771               NaN              NaN  \n",
       "778               NaN              NaN  \n",
       "491               NaN              NaN  \n",
       "788               NaN              NaN  \n",
       "791               NaN              NaN  \n",
       "890               NaN              NaN  \n",
       "493               NaN              NaN  \n",
       "850               NaN              NaN  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm2[\n",
    "    dm2.duplicated(subset=[\"Attribute\", \"Parent\", \"DependsOn\"], keep=False)\n",
    "].sort_values(by=[\"Parent\", \"Attribute\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Attribute, Description, Valid Values, DependsOn, Properties, Required, Parent, DependsOn Component, Source, Validation Rules]\n",
       "Index: []"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm2.query('Attribute == \"UnknownNot collected\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(dm2.shape)\n",
    "\n",
    "# dm2 = dm2.drop_duplicates(\n",
    "#     subset=['Attribute', 'Parent', 'DependsOn']).reset_index(drop=True)\n",
    "\n",
    "# display(dm2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with pd.option_context('display.max_rows', None, 'display.max_colwidth', 0):\n",
    "#     display(dm[dm['Attribute'] == 'sampleType'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check columns for Speical Characters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>reagentID(s)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>treatmentID(s)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>Cardiovascular Health Study (CHS)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>MAXO codes or labels (not listed for purposes of this RFC)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>Study of Osteoporotic Fractures (SOF)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>489</th>\n",
       "      <td>The Osteoporotic Fractures in Men (MrOS) Study</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492</th>\n",
       "      <td>and Body Composition Study (HealthABC)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>CCF-BSO values (not listed for purposes of this RFC)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537</th>\n",
       "      <td>Cerebrospinal Fluid (CSF)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578</th>\n",
       "      <td>Genbank common names (not listed for purposes of this RFC)Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>586</th>\n",
       "      <td>Human Reference Atlas values (not listed for purposes of this RFC)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>606</th>\n",
       "      <td>Illumina Infinium MethylationEPIC BeadChip v1.0 (850k)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>Illumina Infinium MethylationEPIC BeadChip v2.0 (935k)Illumina MiSeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>MassBank of North America (MoNA)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666</th>\n",
       "      <td>NCBI taxonomy values (not listed for purposes of this RFC)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>667</th>\n",
       "      <td>NCI values (not listed for purposes of this RFC)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>World Bank country codes (not listed for purposes of this RFC)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                Attribute\n",
       "144                                                          reagentID(s)\n",
       "152                                                        treatmentID(s)\n",
       "480                                     Cardiovascular Health Study (CHS)\n",
       "484            MAXO codes or labels (not listed for purposes of this RFC)\n",
       "487                                 Study of Osteoporotic Fractures (SOF)\n",
       "489                        The Osteoporotic Fractures in Men (MrOS) Study\n",
       "492                                and Body Composition Study (HealthABC)\n",
       "531                  CCF-BSO values (not listed for purposes of this RFC)\n",
       "537                                             Cerebrospinal Fluid (CSF)\n",
       "578     Genbank common names (not listed for purposes of this RFC)Unknown\n",
       "586    Human Reference Atlas values (not listed for purposes of this RFC)\n",
       "606                Illumina Infinium MethylationEPIC BeadChip v1.0 (850k)\n",
       "607  Illumina Infinium MethylationEPIC BeadChip v2.0 (935k)Illumina MiSeq\n",
       "654                                      MassBank of North America (MoNA)\n",
       "666            NCBI taxonomy values (not listed for purposes of this RFC)\n",
       "667                      NCI values (not listed for purposes of this RFC)\n",
       "799        World Bank country codes (not listed for purposes of this RFC)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "check_cols = [\"Attribute\"]\n",
    "\n",
    "mask = np.column_stack(\n",
    "    [dm2[col].str.contains(\"\\(|\\)\", na=False) for col in dm2[check_cols]]\n",
    ")\n",
    "\n",
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm2[check_cols].loc[mask.any(axis=1)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create Manifests in data model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with pd.option_context('display.max_rows', None, 'display.max_colwidth', 0):\n",
    "#     display(dm2.query('Properties == \"dataProperty\"').groupby(\n",
    "#         'Parent').agg({'Attribute': list}).reset_index())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the manifest templates based on the tables\n",
    "manifests = (\n",
    "    dm2.query('Properties == \"dataProperty\"')\n",
    "    .groupby(\"Parent\")\n",
    "    .agg({\"Attribute\": list})\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "manifests[\"Attribute\"] = manifests[\"Attribute\"].apply(lambda x: x + [\"Component\"])\n",
    "\n",
    "manifests[\"Required\"] = \"True\"\n",
    "manifests[\"Properties\"] = \"dataType\"\n",
    "\n",
    "manifests = manifests.rename({\"Attribute\": \"DependsOn\", \"Parent\": \"Attribute\"}, axis=1)\n",
    "manifests[\"DependsOn\"] = manifests[\"DependsOn\"].apply(lambda x: \",\".join(x))\n",
    "\n",
    "\n",
    "# Store the data types in a list\n",
    "dataTypes = manifests[\"Attribute\"].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(904, 10)\n"
     ]
    }
   ],
   "source": [
    "dm2 = dm2[~dm2[\"Attribute\"].isin(dataTypes)]\n",
    "\n",
    "dm2 = pd.concat([dm2, manifests], axis=0, ignore_index=True)\n",
    "\n",
    "print(dm2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>10x Visium Spatial Gene Expression</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>10x Visium Spatial Gene Expression</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>FALSE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>FALSE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482</th>\n",
       "      <td>FALSE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>825</th>\n",
       "      <td>Lipid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639</th>\n",
       "      <td>Lipid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>690</th>\n",
       "      <td>Not Specified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>Not Specified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>OtherLibraryPreparationMethod</td>\n",
       "      <td>If \"other\" list the name of the library prepar...</td>\n",
       "      <td></td>\n",
       "      <td>specifyLibraryPreparationMethod</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>OtherLibraryPreparationMethod</td>\n",
       "      <td>If \"other\" list the name of the library prepar...</td>\n",
       "      <td></td>\n",
       "      <td>specifyLibraryPreparationMethod</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>OtherLibraryVersion</td>\n",
       "      <td>If \"other\" list the name of the library version</td>\n",
       "      <td></td>\n",
       "      <td>specifyLibraryVersion</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>OtherLibraryVersion</td>\n",
       "      <td>If \"other\" list the name of the library version</td>\n",
       "      <td></td>\n",
       "      <td>specifyLibraryVersion</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>Plasma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>857</th>\n",
       "      <td>Plasma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731</th>\n",
       "      <td>Prefer not to answer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>Prefer not to answer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>734</th>\n",
       "      <td>Protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>Protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749</th>\n",
       "      <td>Saliva</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864</th>\n",
       "      <td>Saliva</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>752</th>\n",
       "      <td>Serum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866</th>\n",
       "      <td>Serum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>760</th>\n",
       "      <td>Sputum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>873</th>\n",
       "      <td>Sputum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771</th>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778</th>\n",
       "      <td>TRUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>491</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>791</th>\n",
       "      <td>Urine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>890</th>\n",
       "      <td>Urine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>dnaBatchID</td>\n",
       "      <td>DNA isolation batch identifier, provided by th...</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not Speci...</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>dnaBatchID</td>\n",
       "      <td>DNA isolation batch</td>\n",
       "      <td>Other,Unknown,Not collected,Not applicable</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>libraryPreparationMethod</td>\n",
       "      <td>Method by which library was prepared</td>\n",
       "      <td>10x,Accel-NGS 2S Plus,Accel-NGS Methyl-Seq,CEL...</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>libraryPreparationMethod</td>\n",
       "      <td>Method by which library was prepared</td>\n",
       "      <td>10x,Accel-NGS 2S Plus,Accel-NGS Methyl-Seq,CEL...</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>libraryVersion</td>\n",
       "      <td>Library Version: for example, rnaSeq 10x libra...</td>\n",
       "      <td>OtherLibraryVersion,Unknown,Not collected,Not ...</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>libraryVersion</td>\n",
       "      <td>Library Version: for example, rnaSeq 10x libra...</td>\n",
       "      <td>OtherLibraryVersion,Unknown,Not collected,Not ...</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>850</th>\n",
       "      <td>pairedEnd</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>pairedEnd</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>specifyLibraryPreparationMethod</td>\n",
       "      <td>If \"other\" list the name of the library prepar...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>specifyLibraryPreparationMethod</td>\n",
       "      <td>If \"other\" list the name of the library prepar...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>specifyLibraryVersion</td>\n",
       "      <td>If \"other\" list the name of the library version</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>specifyLibraryVersion</td>\n",
       "      <td>If \"other\" list the name of the library version</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Attribute  \\\n",
       "497  10x Visium Spatial Gene Expression   \n",
       "495  10x Visium Spatial Gene Expression   \n",
       "568                               FALSE   \n",
       "483                               FALSE   \n",
       "482                               FALSE   \n",
       "825                               Lipid   \n",
       "639                               Lipid   \n",
       "690                       Not Specified   \n",
       "687                       Not Specified   \n",
       "367       OtherLibraryPreparationMethod   \n",
       "370       OtherLibraryPreparationMethod   \n",
       "368                 OtherLibraryVersion   \n",
       "371                 OtherLibraryVersion   \n",
       "721                              Plasma   \n",
       "857                              Plasma   \n",
       "731                Prefer not to answer   \n",
       "486                Prefer not to answer   \n",
       "734                             Protein   \n",
       "859                             Protein   \n",
       "749                              Saliva   \n",
       "864                              Saliva   \n",
       "752                               Serum   \n",
       "866                               Serum   \n",
       "760                              Sputum   \n",
       "873                              Sputum   \n",
       "771                                TRUE   \n",
       "778                                TRUE   \n",
       "788                             Unknown   \n",
       "491                             Unknown   \n",
       "791                               Urine   \n",
       "890                               Urine   \n",
       "10                           dnaBatchID   \n",
       "9                            dnaBatchID   \n",
       "25             libraryPreparationMethod   \n",
       "17             libraryPreparationMethod   \n",
       "27                       libraryVersion   \n",
       "19                       libraryVersion   \n",
       "850                           pairedEnd   \n",
       "493                           pairedEnd   \n",
       "26      specifyLibraryPreparationMethod   \n",
       "18      specifyLibraryPreparationMethod   \n",
       "28                specifyLibraryVersion   \n",
       "20                specifyLibraryVersion   \n",
       "\n",
       "                                           Description  \\\n",
       "497                                                NaN   \n",
       "495                                                NaN   \n",
       "568                                                NaN   \n",
       "483                                                NaN   \n",
       "482                                                NaN   \n",
       "825                                                NaN   \n",
       "639                                                NaN   \n",
       "690                                                NaN   \n",
       "687                                                NaN   \n",
       "367  If \"other\" list the name of the library prepar...   \n",
       "370  If \"other\" list the name of the library prepar...   \n",
       "368    If \"other\" list the name of the library version   \n",
       "371    If \"other\" list the name of the library version   \n",
       "721                                                NaN   \n",
       "857                                                NaN   \n",
       "731                                                NaN   \n",
       "486                                                NaN   \n",
       "734                                                NaN   \n",
       "859                                                NaN   \n",
       "749                                                NaN   \n",
       "864                                                NaN   \n",
       "752                                                NaN   \n",
       "866                                                NaN   \n",
       "760                                                NaN   \n",
       "873                                                NaN   \n",
       "771                                                NaN   \n",
       "778                                                NaN   \n",
       "788                                                NaN   \n",
       "491                                                NaN   \n",
       "791                                                NaN   \n",
       "890                                                NaN   \n",
       "10   DNA isolation batch identifier, provided by th...   \n",
       "9                                  DNA isolation batch   \n",
       "25                Method by which library was prepared   \n",
       "17                Method by which library was prepared   \n",
       "27   Library Version: for example, rnaSeq 10x libra...   \n",
       "19   Library Version: for example, rnaSeq 10x libra...   \n",
       "850                                                NaN   \n",
       "493                                                NaN   \n",
       "26   If \"other\" list the name of the library prepar...   \n",
       "18   If \"other\" list the name of the library prepar...   \n",
       "28     If \"other\" list the name of the library version   \n",
       "20     If \"other\" list the name of the library version   \n",
       "\n",
       "                                          Valid Values  \\\n",
       "497                                                NaN   \n",
       "495                                                NaN   \n",
       "568                                                NaN   \n",
       "483                                                NaN   \n",
       "482                                                NaN   \n",
       "825                                                NaN   \n",
       "639                                                NaN   \n",
       "690                                                NaN   \n",
       "687                                                NaN   \n",
       "367                                                      \n",
       "370                                                      \n",
       "368                                                      \n",
       "371                                                      \n",
       "721                                                NaN   \n",
       "857                                                NaN   \n",
       "731                                                NaN   \n",
       "486                                                NaN   \n",
       "734                                                NaN   \n",
       "859                                                NaN   \n",
       "749                                                NaN   \n",
       "864                                                NaN   \n",
       "752                                                NaN   \n",
       "866                                                NaN   \n",
       "760                                                NaN   \n",
       "873                                                NaN   \n",
       "771                                                NaN   \n",
       "778                                                NaN   \n",
       "788                                                NaN   \n",
       "491                                                NaN   \n",
       "791                                                NaN   \n",
       "890                                                NaN   \n",
       "10   Unknown,Not collected,Not applicable,Not Speci...   \n",
       "9           Other,Unknown,Not collected,Not applicable   \n",
       "25   10x,Accel-NGS 2S Plus,Accel-NGS Methyl-Seq,CEL...   \n",
       "17   10x,Accel-NGS 2S Plus,Accel-NGS Methyl-Seq,CEL...   \n",
       "27   OtherLibraryVersion,Unknown,Not collected,Not ...   \n",
       "19   OtherLibraryVersion,Unknown,Not collected,Not ...   \n",
       "850                                                NaN   \n",
       "493                                                NaN   \n",
       "26                                                       \n",
       "18                                                       \n",
       "28                                                       \n",
       "20                                                       \n",
       "\n",
       "                           DependsOn    Properties Required  \\\n",
       "497                              NaN    validValue    False   \n",
       "495                              NaN    validValue    False   \n",
       "568                              NaN    validValue    False   \n",
       "483                              NaN    validValue    False   \n",
       "482                              NaN    validValue    False   \n",
       "825                              NaN    validValue    False   \n",
       "639                              NaN    validValue    False   \n",
       "690                              NaN    validValue    False   \n",
       "687                              NaN    validValue    False   \n",
       "367  specifyLibraryPreparationMethod    validValue    False   \n",
       "370  specifyLibraryPreparationMethod    validValue    False   \n",
       "368            specifyLibraryVersion    validValue    False   \n",
       "371            specifyLibraryVersion    validValue    False   \n",
       "721                              NaN    validValue    False   \n",
       "857                              NaN    validValue    False   \n",
       "731                              NaN    validValue    False   \n",
       "486                              NaN    validValue    False   \n",
       "734                              NaN    validValue    False   \n",
       "859                              NaN    validValue    False   \n",
       "749                              NaN    validValue    False   \n",
       "864                              NaN    validValue    False   \n",
       "752                              NaN    validValue    False   \n",
       "866                              NaN    validValue    False   \n",
       "760                              NaN    validValue    False   \n",
       "873                              NaN    validValue    False   \n",
       "771                              NaN    validValue    False   \n",
       "778                              NaN    validValue    False   \n",
       "788                              NaN    validValue    False   \n",
       "491                              NaN    validValue    False   \n",
       "791                              NaN    validValue    False   \n",
       "890                              NaN    validValue    False   \n",
       "10                                    dataProperty     True   \n",
       "9                                     dataProperty     True   \n",
       "25                                    dataProperty     True   \n",
       "17                                    dataProperty     True   \n",
       "27                                    dataProperty     True   \n",
       "19                                    dataProperty     True   \n",
       "850                              NaN    validValue    False   \n",
       "493                              NaN    validValue    False   \n",
       "26                                    dataProperty    False   \n",
       "18                                    dataProperty    False   \n",
       "28                                    dataProperty    False   \n",
       "20                                    dataProperty    False   \n",
       "\n",
       "                                               Parent DependsOn Component  \\\n",
       "497                                               NaN                 NaN   \n",
       "495                                               NaN                 NaN   \n",
       "568                                               NaN                 NaN   \n",
       "483                                               NaN                 NaN   \n",
       "482                                               NaN                 NaN   \n",
       "825                                               NaN                 NaN   \n",
       "639                                               NaN                 NaN   \n",
       "690                                               NaN                 NaN   \n",
       "687                                               NaN                 NaN   \n",
       "367  bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "370  bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "368  bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "371  bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "721                                               NaN                 NaN   \n",
       "857                                               NaN                 NaN   \n",
       "731                                               NaN                 NaN   \n",
       "486                                               NaN                 NaN   \n",
       "734                                               NaN                 NaN   \n",
       "859                                               NaN                 NaN   \n",
       "749                                               NaN                 NaN   \n",
       "864                                               NaN                 NaN   \n",
       "752                                               NaN                 NaN   \n",
       "866                                               NaN                 NaN   \n",
       "760                                               NaN                 NaN   \n",
       "873                                               NaN                 NaN   \n",
       "771                                               NaN                 NaN   \n",
       "778                                               NaN                 NaN   \n",
       "788                                               NaN                 NaN   \n",
       "491                                               NaN                 NaN   \n",
       "791                                               NaN                 NaN   \n",
       "890                                               NaN                 NaN   \n",
       "10   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "9    bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "25   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "17   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "27   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "19   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "850                                               NaN                 NaN   \n",
       "493                                               NaN                 NaN   \n",
       "26   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "18   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "28   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "20   bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "\n",
       "               Source Validation Rules  \n",
       "497               NaN              NaN  \n",
       "495               NaN              NaN  \n",
       "568               NaN              NaN  \n",
       "483               NaN              NaN  \n",
       "482               NaN              NaN  \n",
       "825               NaN              NaN  \n",
       "639               NaN              NaN  \n",
       "690               NaN              NaN  \n",
       "687               NaN              NaN  \n",
       "367  Sage Bionetworks                   \n",
       "370  Sage Bionetworks                   \n",
       "368  Sage Bionetworks                   \n",
       "371  Sage Bionetworks                   \n",
       "721               NaN              NaN  \n",
       "857               NaN              NaN  \n",
       "731               NaN              NaN  \n",
       "486               NaN              NaN  \n",
       "734               NaN              NaN  \n",
       "859               NaN              NaN  \n",
       "749               NaN              NaN  \n",
       "864               NaN              NaN  \n",
       "752               NaN              NaN  \n",
       "866               NaN              NaN  \n",
       "760               NaN              NaN  \n",
       "873               NaN              NaN  \n",
       "771               NaN              NaN  \n",
       "778               NaN              NaN  \n",
       "788               NaN              NaN  \n",
       "491               NaN              NaN  \n",
       "791               NaN              NaN  \n",
       "890               NaN              NaN  \n",
       "10   Sage Bionetworks                   \n",
       "9    Sage Bionetworks                   \n",
       "25   Sage Bionetworks                   \n",
       "17   Sage Bionetworks                   \n",
       "27   Sage Bionetworks                   \n",
       "19   Sage Bionetworks                   \n",
       "850               NaN              NaN  \n",
       "493               NaN              NaN  \n",
       "26   Sage Bionetworks                   \n",
       "18   Sage Bionetworks                   \n",
       "28   Sage Bionetworks                   \n",
       "20   Sage Bionetworks                   "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm2[\n",
    "    dm2.duplicated(\n",
    "        subset=[\"Attribute\", \"DependsOn\", \"Properties\", \"Parent\"], keep=False\n",
    "    )\n",
    "].sort_values(by=\"Attribute\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Another typo\n",
    "dm2[\"Valid Values\"] = dm2[\"Valid Values\"].str.replace(\n",
    "    \"falseFalseFALSEtrueTrueTRUE\", \"TRUE,FALSE\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['specimenID',\n",
       " 'sampleType',\n",
       " 'specifySampleType',\n",
       " 'specifyMeasurementTechnique',\n",
       " 'technologyPlatformVersion',\n",
       " 'specifyPlatformVersion',\n",
       " 'platformLocation',\n",
       " 'specifyPlatformLocation',\n",
       " 'referenceTranscriptID',\n",
       " 'repositoryName',\n",
       " 'specifyRepositoryName',\n",
       " 'transcriptType',\n",
       " 'specifyTranscriptType',\n",
       " 'resultUnit',\n",
       " 'valueReported',\n",
       " 'libraryPreparationMethod',\n",
       " 'specifyLibraryPreparationMethod',\n",
       " 'libraryVersion',\n",
       " 'specifyLibraryVersion',\n",
       " 'isStranded',\n",
       " 'readStrandOrigin',\n",
       " 'readLength',\n",
       " 'readLengthUnits',\n",
       " 'specifyReadLengthUnits',\n",
       " 'runType',\n",
       " 'totalReads',\n",
       " 'Component']"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm2[dm2[\"Attribute\"] == \"RNAseq\"][\"DependsOn\"].values[0].split(\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = dm2.drop_duplicates(subset=[\"Attribute\", \"Parent\"]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['specimenID',\n",
       " 'sampleType',\n",
       " 'specifySampleType',\n",
       " 'specifyMeasurementTechnique',\n",
       " 'technologyPlatformVersion',\n",
       " 'specifyPlatformVersion',\n",
       " 'platformLocation',\n",
       " 'specifyPlatformLocation',\n",
       " 'referenceTranscriptID',\n",
       " 'repositoryName',\n",
       " 'specifyRepositoryName',\n",
       " 'transcriptType',\n",
       " 'specifyTranscriptType',\n",
       " 'resultUnit',\n",
       " 'valueReported',\n",
       " 'libraryPreparationMethod',\n",
       " 'specifyLibraryPreparationMethod',\n",
       " 'libraryVersion',\n",
       " 'specifyLibraryVersion',\n",
       " 'isStranded',\n",
       " 'readStrandOrigin',\n",
       " 'readLength',\n",
       " 'readLengthUnits',\n",
       " 'specifyReadLengthUnits',\n",
       " 'runType',\n",
       " 'totalReads',\n",
       " 'Component']"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp[temp[\"Attribute\"] == \"RNAseq\"][\"DependsOn\"].values[0].split(\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(904, 10)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(632, 10)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(dm2.shape)\n",
    "\n",
    "dm2 = dm2.drop_duplicates(subset=[\"Attribute\"]).reset_index(\n",
    "    drop=True\n",
    ")  # 'DependsOn', 'Properties',\n",
    "\n",
    "display(dm2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Attribute, Description, Valid Values, DependsOn, Properties, Required, Parent, DependsOn Component, Source, Validation Rules]\n",
       "Index: []"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Recode required columns and fix spelling mistakes\n",
    "required_recoder = {\"0.0\": \"False\", \"1.0\": \"True\", \"FASLSE\": \"False\"}\n",
    "dm2[\"Required\"] = dm2[\"Required\"].replace(required_recoder)\n",
    "\n",
    "dm2[~dm2[\"Required\"].str.contains(\"True|False\", regex=True)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the measurement technique used</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>OtherMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the measurement technique used</td>\n",
       "      <td></td>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>OtherMetabolomics HumanmeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the measurement technique used</td>\n",
       "      <td></td>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,msTarget,msAnalyzerTypeMS1,specifyMSAnalyzerTypeMS1,msAnalyzerTypeMS2,specifyMSAnalyzerTypeMS2,msAssayTechnique,specifyMSAssayTechnique,msAnalyteType,specifyMSAnalyteType,msInstrumentModel,specifyMSInstrumentModel,sampleIntroduction,specifySampleIntroduction,extractionMethod,acquisitionMode,acquisitionSoftware,acquistionSoftwareVersion,samplePrepProtocol,databaseName,specifyDatabaseName,databaseSource,specifyDatabaseSource,databaseWeblink,spectrometerFrequency,hasIonizationSource,ionProperty,vacuumPressure,vacuumPressureUnit,specifyVacuumPressueUnits,lensVoltages ,lensVoltagesUnit ,specifyLensVoltageUnits,gasFlowTemperature,gasFlowTemperatureUnit,hasAssayControl,controlType,specifyControlType,experiementalBatchSize,experiementalBatchSizeUnit,specifyExperimentalBatchSizeUnits,batchSize,batchSizeUnit,specifyBatchSizeUnits,batchID,batchLabel,sampleBatchID,sampleBatchSize,sampleBatchSizeUnit,specifySampleBatchSizeUnits,acquisitionBatchID,acquisitionBatchSize,acquisitionBatchSizeUnit,specifyAcquisitionBatchSizeUnits,processingBatchID,processingBatchSize,processingBatchSizeUnit,specifyProcessingBatchSizeUnit,f,Component</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>Microbiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,extractionMethod,specifyExtractionMethod,libraryBatchID,sequencingBatchID,sequencingBatchSize,sequencingBatchSizeUnit,specifySequencingBatchSizeUnit,dnaBatchID,dnaBatchSize,dnaBatchSizeUnit,specifyDNABatchSizeUnit,libraryPrep,specifyLibraryPrep,libraryPreparationMethod,specifyLibraryPreparationMethod,libraryVersion,specifyLibraryVersion,isStranded,readStrandOrigin,readLength,readLengthUnits,specifyReadLengthUnits,runType,totalReads,Component</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>626</th>\n",
       "      <td>RNAseq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,referenceTranscriptID,repositoryName,specifyRepositoryName,transcriptType,specifyTranscriptType,resultUnit,valueReported,libraryPreparationMethod,specifyLibraryPreparationMethod,libraryVersion,specifyLibraryVersion,isStranded,readStrandOrigin,readLength,readLengthUnits,specifyReadLengthUnits,runType,totalReads,Component</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,referenceTranscriptID,repositoryName,specifyRepositoryName,transcriptType,specifyTranscriptType,resultUnit,valueReported,rnaBatchID,libraryBatchID,sequencingBatchID,libraryPrep,specifyLibraryPrep,libraryPreparationMethod,specifyLibraryPreparationMethod,libraryVersion,specifyLibraryVersion,isStranded,readStrandOrigin,readLength,readLengthUnits,specifyReadLengthUnits,runType,totalReads,Component</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,referenceTranscriptID,dnaBatchID,dnaBatchID,dnaBatchSize,dnaBatchSizeUnit,specifyDNABatchSizeUnit,libraryBatchID,libraryPrep,specifyLibraryPrep,libraryPreparationMethod,specifyLibraryPreparationMethod,libraryVersion,specifyLibraryVersion,sequencingBatchID,sequencingBatchSize,sequencingBatchSizeUnit,specifySequencingBatchSizeUnit,libraryPreparationMethod,specifyLibraryPreparationMethod,libraryVersion,specifyLibraryVersion,isStranded,readStrandOrigin,readLength,readLengthUnits,specifyReadLengthUnits,runType,totalReads,directionalBSseqLibrary,gDNAconc,lambdaDNAconc,pcrCycles,meanCoverage,conversionRatio,conversionRatioUnits,specifyConversionRatioUnit,enrichmentMethod,specifyEnrichmentMethod,Component</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>genotyping</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,useReagent,reagentID(s),reagentName,reagentManufacturer,reagentCatalogNumber,reagentLotNumber,reagentWeblink,reagentContact,useTreatment,treatmentID(s),treatmentName,treatmentAmountValue,treatmentAmountUnit,treatmentDurationValue,treatmentDurationUnit,treatmentTemperatureValue,treatmentTemperatureUnit,specifyMeasurementTechnique,Component</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,msTarget,msAnalyzerTypeMS1,specifyMSAnalyzerTypeMS1,msAnalyzerTypeMS2,specifyMSAnalyzerTypeMS2,msAssayTechnique,specifyMSAssayTechnique,msAnalyteType,specifyMSAnalyteType,protoemicsAssayType,specifyProteomicsAssayType,msInstrumentModel,specifyMSInstrumentModel,digestionMethod,specifyDigestionMethod,fractionIdentifier,acquisitionMode,acquisitionSoftware,acquistionSoftwareVersion,samplePrepProtocol,databaseName,specifyDatabaseName,databaseSource,specifyDatabaseSource,databaseWeblink,spectrometerFrequency,modificationParameters,specifyOtherModificationParameters,cleavageAgents,specifyOtherCleavageAgent,fragmentMassTolerance,precursorMassTolerance,dataFile,Component</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,referenceTranscriptID,repositoryName,specifyRepositoryName,transcriptType,specifyTranscriptType,resultUnit,valueReported,rnaBatchID,rnaBatchSize,rnaBatchSizeUnit,specifyRNABatchSizeUnit,libraryBatchID,sequencingBatchID,sequencingBatchSize,sequencingBatchSizeUnit,specifySequencingBatchSizeUnit,libraryPrep,specifyLibraryPrep,libraryPreparationMethod,specifyLibraryPreparationMethod,libraryVersion,specifyLibraryVersion,isStranded,readStrandOrigin,readLength,readLengthUnits,specifyReadLengthUnits,runType,totalReads,Component</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            Attribute  \\\n",
       "3                         specifyMeasurementTechnique   \n",
       "180                         OtherMeasurementTechnique   \n",
       "194       OtherMetabolomics HumanmeasurementTechnique   \n",
       "624                                Metabolomics Human   \n",
       "625                                        Microbiome   \n",
       "626                                            RNAseq   \n",
       "627                           Whole Genome Sequencing   \n",
       "628  bsSeq (bisulfite-seq WGBS methylseq methylomics)   \n",
       "629                                        genotyping   \n",
       "630                                        proteomics   \n",
       "631                                          scRNAseq   \n",
       "\n",
       "                                                                Description  \\\n",
       "3    If \"other\" is selected list the name of the measurement technique used   \n",
       "180  If \"other\" is selected list the name of the measurement technique used   \n",
       "194  If \"other\" is selected list the name of the measurement technique used   \n",
       "624                                                                     NaN   \n",
       "625                                                                     NaN   \n",
       "626                                                                     NaN   \n",
       "627                                                                     NaN   \n",
       "628                                                                     NaN   \n",
       "629                                                                     NaN   \n",
       "630                                                                     NaN   \n",
       "631                                                                     NaN   \n",
       "\n",
       "    Valid Values  \\\n",
       "3                  \n",
       "180                \n",
       "194                \n",
       "624          NaN   \n",
       "625          NaN   \n",
       "626          NaN   \n",
       "627          NaN   \n",
       "628          NaN   \n",
       "629          NaN   \n",
       "630          NaN   \n",
       "631          NaN   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     DependsOn  \\\n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                \n",
       "180                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                specifyMeasurementTechnique   \n",
       "194                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                specifyMeasurementTechnique   \n",
       "624  specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,msTarget,msAnalyzerTypeMS1,specifyMSAnalyzerTypeMS1,msAnalyzerTypeMS2,specifyMSAnalyzerTypeMS2,msAssayTechnique,specifyMSAssayTechnique,msAnalyteType,specifyMSAnalyteType,msInstrumentModel,specifyMSInstrumentModel,sampleIntroduction,specifySampleIntroduction,extractionMethod,acquisitionMode,acquisitionSoftware,acquistionSoftwareVersion,samplePrepProtocol,databaseName,specifyDatabaseName,databaseSource,specifyDatabaseSource,databaseWeblink,spectrometerFrequency,hasIonizationSource,ionProperty,vacuumPressure,vacuumPressureUnit,specifyVacuumPressueUnits,lensVoltages ,lensVoltagesUnit ,specifyLensVoltageUnits,gasFlowTemperature,gasFlowTemperatureUnit,hasAssayControl,controlType,specifyControlType,experiementalBatchSize,experiementalBatchSizeUnit,specifyExperimentalBatchSizeUnits,batchSize,batchSizeUnit,specifyBatchSizeUnits,batchID,batchLabel,sampleBatchID,sampleBatchSize,sampleBatchSizeUnit,specifySampleBatchSizeUnits,acquisitionBatchID,acquisitionBatchSize,acquisitionBatchSizeUnit,specifyAcquisitionBatchSizeUnits,processingBatchID,processingBatchSize,processingBatchSizeUnit,specifyProcessingBatchSizeUnit,f,Component   \n",
       "625                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,extractionMethod,specifyExtractionMethod,libraryBatchID,sequencingBatchID,sequencingBatchSize,sequencingBatchSizeUnit,specifySequencingBatchSizeUnit,dnaBatchID,dnaBatchSize,dnaBatchSizeUnit,specifyDNABatchSizeUnit,libraryPrep,specifyLibraryPrep,libraryPreparationMethod,specifyLibraryPreparationMethod,libraryVersion,specifyLibraryVersion,isStranded,readStrandOrigin,readLength,readLengthUnits,specifyReadLengthUnits,runType,totalReads,Component   \n",
       "626                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,referenceTranscriptID,repositoryName,specifyRepositoryName,transcriptType,specifyTranscriptType,resultUnit,valueReported,libraryPreparationMethod,specifyLibraryPreparationMethod,libraryVersion,specifyLibraryVersion,isStranded,readStrandOrigin,readLength,readLengthUnits,specifyReadLengthUnits,runType,totalReads,Component   \n",
       "627                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,referenceTranscriptID,repositoryName,specifyRepositoryName,transcriptType,specifyTranscriptType,resultUnit,valueReported,rnaBatchID,libraryBatchID,sequencingBatchID,libraryPrep,specifyLibraryPrep,libraryPreparationMethod,specifyLibraryPreparationMethod,libraryVersion,specifyLibraryVersion,isStranded,readStrandOrigin,readLength,readLengthUnits,specifyReadLengthUnits,runType,totalReads,Component   \n",
       "628                                                                                                                                                                                                                                                                                                                                                                                                                                           specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,referenceTranscriptID,dnaBatchID,dnaBatchID,dnaBatchSize,dnaBatchSizeUnit,specifyDNABatchSizeUnit,libraryBatchID,libraryPrep,specifyLibraryPrep,libraryPreparationMethod,specifyLibraryPreparationMethod,libraryVersion,specifyLibraryVersion,sequencingBatchID,sequencingBatchSize,sequencingBatchSizeUnit,specifySequencingBatchSizeUnit,libraryPreparationMethod,specifyLibraryPreparationMethod,libraryVersion,specifyLibraryVersion,isStranded,readStrandOrigin,readLength,readLengthUnits,specifyReadLengthUnits,runType,totalReads,directionalBSseqLibrary,gDNAconc,lambdaDNAconc,pcrCycles,meanCoverage,conversionRatio,conversionRatioUnits,specifyConversionRatioUnit,enrichmentMethod,specifyEnrichmentMethod,Component   \n",
       "629                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               specimenID,sampleType,specifySampleType,useReagent,reagentID(s),reagentName,reagentManufacturer,reagentCatalogNumber,reagentLotNumber,reagentWeblink,reagentContact,useTreatment,treatmentID(s),treatmentName,treatmentAmountValue,treatmentAmountUnit,treatmentDurationValue,treatmentDurationUnit,treatmentTemperatureValue,treatmentTemperatureUnit,specifyMeasurementTechnique,Component   \n",
       "630                                                                                                                                                                                                                                                                                                                                                                                                                                                                               specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,msTarget,msAnalyzerTypeMS1,specifyMSAnalyzerTypeMS1,msAnalyzerTypeMS2,specifyMSAnalyzerTypeMS2,msAssayTechnique,specifyMSAssayTechnique,msAnalyteType,specifyMSAnalyteType,protoemicsAssayType,specifyProteomicsAssayType,msInstrumentModel,specifyMSInstrumentModel,digestionMethod,specifyDigestionMethod,fractionIdentifier,acquisitionMode,acquisitionSoftware,acquistionSoftwareVersion,samplePrepProtocol,databaseName,specifyDatabaseName,databaseSource,specifyDatabaseSource,databaseWeblink,spectrometerFrequency,modificationParameters,specifyOtherModificationParameters,cleavageAgents,specifyOtherCleavageAgent,fragmentMassTolerance,precursorMassTolerance,dataFile,Component   \n",
       "631                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                specimenID,sampleType,specifySampleType,specifyMeasurementTechnique,technologyPlatformVersion,specifyPlatformVersion,platformLocation,specifyPlatformLocation,referenceTranscriptID,repositoryName,specifyRepositoryName,transcriptType,specifyTranscriptType,resultUnit,valueReported,rnaBatchID,rnaBatchSize,rnaBatchSizeUnit,specifyRNABatchSizeUnit,libraryBatchID,sequencingBatchID,sequencingBatchSize,sequencingBatchSizeUnit,specifySequencingBatchSizeUnit,libraryPrep,specifyLibraryPrep,libraryPreparationMethod,specifyLibraryPreparationMethod,libraryVersion,specifyLibraryVersion,isStranded,readStrandOrigin,readLength,readLengthUnits,specifyReadLengthUnits,runType,totalReads,Component   \n",
       "\n",
       "       Properties Required                                            Parent  \\\n",
       "3    dataProperty    False  bsSeq (bisulfite-seq WGBS methylseq methylomics)   \n",
       "180    validValue    False  bsSeq (bisulfite-seq WGBS methylseq methylomics)   \n",
       "194    validValue    False                                Metabolomics Human   \n",
       "624      dataType     True                                               NaN   \n",
       "625      dataType     True                                               NaN   \n",
       "626      dataType     True                                               NaN   \n",
       "627      dataType     True                                               NaN   \n",
       "628      dataType     True                                               NaN   \n",
       "629      dataType     True                                               NaN   \n",
       "630      dataType     True                                               NaN   \n",
       "631      dataType     True                                               NaN   \n",
       "\n",
       "    DependsOn Component            Source Validation Rules  \n",
       "3                        Sage Bionetworks                   \n",
       "180                      Sage Bionetworks                   \n",
       "194                      Sage Bionetworks                   \n",
       "624                 NaN               NaN              NaN  \n",
       "625                 NaN               NaN              NaN  \n",
       "626                 NaN               NaN              NaN  \n",
       "627                 NaN               NaN              NaN  \n",
       "628                 NaN               NaN              NaN  \n",
       "629                 NaN               NaN              NaN  \n",
       "630                 NaN               NaN              NaN  \n",
       "631                 NaN               NaN              NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>OtherMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>OtherMetabolomics HumanmeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>validValue</td>\n",
       "      <td>False</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specif...</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>Microbiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specif...</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>626</th>\n",
       "      <td>RNAseq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specif...</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specif...</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>bsSeq (bisulfite-seq WGBS methylseq methylomics)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specif...</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>genotyping</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,useRea...</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specif...</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specimenID,sampleType,specifySampleType,specif...</td>\n",
       "      <td>dataType</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            Attribute  \\\n",
       "3                         specifyMeasurementTechnique   \n",
       "180                         OtherMeasurementTechnique   \n",
       "194       OtherMetabolomics HumanmeasurementTechnique   \n",
       "624                                Metabolomics Human   \n",
       "625                                        Microbiome   \n",
       "626                                            RNAseq   \n",
       "627                           Whole Genome Sequencing   \n",
       "628  bsSeq (bisulfite-seq WGBS methylseq methylomics)   \n",
       "629                                        genotyping   \n",
       "630                                        proteomics   \n",
       "631                                          scRNAseq   \n",
       "\n",
       "                                           Description Valid Values  \\\n",
       "3    If \"other\" is selected list the name of the me...                \n",
       "180  If \"other\" is selected list the name of the me...                \n",
       "194  If \"other\" is selected list the name of the me...                \n",
       "624                                                NaN          NaN   \n",
       "625                                                NaN          NaN   \n",
       "626                                                NaN          NaN   \n",
       "627                                                NaN          NaN   \n",
       "628                                                NaN          NaN   \n",
       "629                                                NaN          NaN   \n",
       "630                                                NaN          NaN   \n",
       "631                                                NaN          NaN   \n",
       "\n",
       "                                             DependsOn    Properties Required  \\\n",
       "3                                                       dataProperty    False   \n",
       "180                        specifyMeasurementTechnique    validValue    False   \n",
       "194                        specifyMeasurementTechnique    validValue    False   \n",
       "624  specimenID,sampleType,specifySampleType,specif...      dataType     True   \n",
       "625  specimenID,sampleType,specifySampleType,specif...      dataType     True   \n",
       "626  specimenID,sampleType,specifySampleType,specif...      dataType     True   \n",
       "627  specimenID,sampleType,specifySampleType,specif...      dataType     True   \n",
       "628  specimenID,sampleType,specifySampleType,specif...      dataType     True   \n",
       "629  specimenID,sampleType,specifySampleType,useRea...      dataType     True   \n",
       "630  specimenID,sampleType,specifySampleType,specif...      dataType     True   \n",
       "631  specimenID,sampleType,specifySampleType,specif...      dataType     True   \n",
       "\n",
       "                                               Parent DependsOn Component  \\\n",
       "3    bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "180  bsSeq (bisulfite-seq WGBS methylseq methylomics)                       \n",
       "194                                Metabolomics Human                       \n",
       "624                                               NaN                 NaN   \n",
       "625                                               NaN                 NaN   \n",
       "626                                               NaN                 NaN   \n",
       "627                                               NaN                 NaN   \n",
       "628                                               NaN                 NaN   \n",
       "629                                               NaN                 NaN   \n",
       "630                                               NaN                 NaN   \n",
       "631                                               NaN                 NaN   \n",
       "\n",
       "               Source Validation Rules  \n",
       "3    Sage Bionetworks                   \n",
       "180  Sage Bionetworks                   \n",
       "194  Sage Bionetworks                   \n",
       "624               NaN              NaN  \n",
       "625               NaN              NaN  \n",
       "626               NaN              NaN  \n",
       "627               NaN              NaN  \n",
       "628               NaN              NaN  \n",
       "629               NaN              NaN  \n",
       "630               NaN              NaN  \n",
       "631               NaN              NaN  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "search_df(dm2, \"measurementTechnique\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Last bit of cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove measurement technique dependency from biospecimen human\n",
    "dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"] = (\n",
    "    dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"]\n",
    "    .values[0]\n",
    "    .replace(\"measurementTechnique,\", \"\")\n",
    ")\n",
    "bio_measure_technique_index = dm2.query(\n",
    "    'Attribute == \"measurementTechnique\" and Parent == \"Biospecimen human\"'\n",
    ")\n",
    "\n",
    "if len(bio_measure_technique_index.index) > 0:\n",
    "    dm2 = dm2.drop(index=bio_measure_technique_index.index[0]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove measurement technique dependency from biospecimen human\n",
    "dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"] = (\n",
    "    dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"]\n",
    "    .values[0]\n",
    "    .replace(\"specifyMeasurementTechnique,\", \"\")\n",
    ")\n",
    "bio_measure_technique_index = dm2.query(\n",
    "    'Attribute == \"specifyMeasurementTechnique\" and Parent == \"Biospecimen human\"'\n",
    ")\n",
    "\n",
    "\n",
    "if len(bio_measure_technique_index.index) > 0:\n",
    "    dm2 = dm2.drop(index=bio_measure_technique_index.index[0]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove measurement technique dependency from biospecimen human\n",
    "dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"] = (\n",
    "    dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"]\n",
    "    .values[0]\n",
    "    .replace(\"OtherMeasurementTechnique,\", \"\")\n",
    ")\n",
    "bio_measure_technique_index = dm2.query('Attribute == \"OtherMeasurementTechnique\"')\n",
    "\n",
    "\n",
    "if len(bio_measure_technique_index.index) > 0:\n",
    "    dm2 = dm2.drop(index=bio_measure_technique_index.index[0]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm2.loc[\n",
    "    dm2[dm2[\"Attribute\"] == \"visitCode\"].index.values[0], \"Valid Values\"\n",
    "] = \"1,2,Other,Unknown,Not collected,Not applicable\"\n",
    "\n",
    "dm2.loc[dm2[dm2[\"Attribute\"] == \"visitCode\"].index.values[0], \"Validation Rules\"] = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extra comma at beginning of valid values\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validation Rules\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixed_attrs = [\n",
    "    {\"attribute\": \"tissueWeight\", \"val_type\": \"mixed float\", \"regex\": \"regex search\"},\n",
    "    {\"attribute\": \"tissueVolume\", \"val_type\": \"mixed float\", \"regex\": \"regex search\"},\n",
    "    {\"attribute\": \"specimenAge\", \"val_type\": \"mixed integer\", \"regex\": \"regex search\"},\n",
    "    {\"attribute\": \"samplingAge\", \"val_type\": \"mixed integer\", \"regex\": \"regex search\"},\n",
    "    {\"attribute\": \"age\", \"val_type\": \"mixed integer\", \"regex\": \"regex search\"},\n",
    "]\n",
    "\n",
    "for ma in mixed_attrs:\n",
    "    attribute = ma[\"attribute\"]\n",
    "    val_type = ma[\"val_type\"]\n",
    "\n",
    "    # get indexes for new validation rules based on attribute\n",
    "    indexes = dm2[dm2[\"Attribute\"] == attribute].index.tolist()\n",
    "\n",
    "    for i in indexes:\n",
    "        if val_type == \"integer\":\n",
    "            first_part = \"[0-9]+\"\n",
    "        elif val_type == \"float\":\n",
    "            first_part = \"^\\d*?\\.?\\d$\"\n",
    "        elif val_type == \"mixed integer\":\n",
    "            regex = \"regex search\"\n",
    "            num_match = \"^\\d*?\"\n",
    "        elif val_type == \"mixed float\":\n",
    "            regex = \"regex search\"\n",
    "            num_match = \"^\\d*?\\.?\\d$\"\n",
    "            # All valid values are applicable\n",
    "            new_string = (\n",
    "                regex\n",
    "                + num_match\n",
    "                + \"|\"\n",
    "                + \"\"\n",
    "                + \"|\".join(dm2.loc[i, \"Valid Values\"].split(\",\"))\n",
    "            )\n",
    "\n",
    "        dm2.loc[i, \"Validation Rules\"] = new_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['',\n",
       " 'regex search ([0-9]+\\\\.[0-9]*.)|([0-9]+)',\n",
       " nan,\n",
       " 'regex search^\\\\d*?\\\\.?\\\\d$|Unknown|Not collected|Not applicable',\n",
       " 'regex search ([0-9]+)']"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm2[\"Validation Rules\"].unique().tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Building Dependencies\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "dependencies = {\n",
    "    \"specimenID\": \"matchAtLeastOne Biospecimenhuman.specimenID value\",\n",
    "    \"individualID\": \"matchExactlyOne IndividualHuman.individualID set\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k, v in dependencies.items():\n",
    "    indexes = dm2[dm2[\"Attribute\"] == k][\"Validation Rules\"].index.values\n",
    "    dm2.loc[indexes, \"Validation Rules\"] = v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm2.loc[3, \"Valid Values\"] = dm2.loc[3, \"Valid Values\"].replace(\n",
    "    \",Whole Genome Sequencing\", \"\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create File Annotations Attributes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "# base file annotations\n",
    "base_file_annotations = {\"resourceType\": \"\", \"isReleased\": False, \"fileType\": \"\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>visitCode</td>\n",
       "      <td>Indicate which longitudinal visit for the individual the data comes from, provided by the data contributor's data dictionary</td>\n",
       "      <td>1,2,Other,Unknown,Not collected,Not applicable</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Attribute  \\\n",
       "161  visitCode   \n",
       "\n",
       "                                                                                                                      Description  \\\n",
       "161  Indicate which longitudinal visit for the individual the data comes from, provided by the data contributor's data dictionary   \n",
       "\n",
       "                                       Valid Values DependsOn    Properties  \\\n",
       "161  1,2,Other,Unknown,Not collected,Not applicable            dataProperty   \n",
       "\n",
       "    Required            Parent DependsOn Component            Source  \\\n",
       "161     True  Individual Human                      Sage Bionetworks   \n",
       "\n",
       "    Validation Rules  \n",
       "161                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm2.query('Attribute == \"visitCode\"'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write out new data model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "hard_coded_valid_values = [\n",
    "    {\n",
    "        \"attribute\": \"visitCode\",\n",
    "        \"valid_value\": \"1,2,Other,Unknown,Not collected,Not applicable\",\n",
    "    },\n",
    "    {\"attribute\": \"consentGroupID\", \"valid_value\": \"1,2,3\"},\n",
    "    {\"attribute\": \"samplingAge\", \"valid_value\": \"\"},\n",
    "    {\"attribute\": \"specimenAge\", \"valid_value\": \"\"},\n",
    "    {\"attribute\": \"age\", \"valid_value\": \"\"},\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "for h in hard_coded_valid_values:\n",
    "    dm2.loc[\n",
    "        dm2[dm2[\"Attribute\"] == h[\"attribute\"]].index.values[0], \"Valid Values\"\n",
    "    ] = h[\"valid_value\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>specimenAge</td>\n",
       "      <td>The subject's age at the time of specimen extraction is the individual's age (since birth) at the time a given specimen was extracted. Measured as age in years.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>Biospecimen human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>regex search^\\d*?\\.?\\d$|Unknown|Not collected|Not applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>samplingAge</td>\n",
       "      <td>The calculated age of the sample, measurement is determined or coded by the data contributor.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>Biospecimen human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>regex search^\\d*?\\.?\\d$|Unknown|Not collected|Not applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>visitCode</td>\n",
       "      <td>Indicate which longitudinal visit for the individual the data comes from, provided by the data contributor's data dictionary</td>\n",
       "      <td>1,2,Other,Unknown,Not collected,Not applicable</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>consentGroupID</td>\n",
       "      <td>Indicate the consent group for the individual, provided by the data contributor's data dictionary</td>\n",
       "      <td>1,2,3</td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>age</td>\n",
       "      <td>Age of the individual (age in years of the individual at first recorded study event (enrollment, visit, observation, sample collection, survey completion, etc.)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>dataProperty</td>\n",
       "      <td>True</td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks,</td>\n",
       "      <td>regex search^\\d*?\\.?\\d$|Unknown|Not collected|Not applicable</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Attribute  \\\n",
       "141     specimenAge   \n",
       "142     samplingAge   \n",
       "161       visitCode   \n",
       "163  consentGroupID   \n",
       "169             age   \n",
       "\n",
       "                                                                                                                                                          Description  \\\n",
       "141  The subject's age at the time of specimen extraction is the individual's age (since birth) at the time a given specimen was extracted. Measured as age in years.   \n",
       "142                                                                    The calculated age of the sample, measurement is determined or coded by the data contributor.    \n",
       "161                                      Indicate which longitudinal visit for the individual the data comes from, provided by the data contributor's data dictionary   \n",
       "163                                                                 Indicate the consent group for the individual, provided by the data contributor's data dictionary   \n",
       "169  Age of the individual (age in years of the individual at first recorded study event (enrollment, visit, observation, sample collection, survey completion, etc.)   \n",
       "\n",
       "                                       Valid Values DependsOn    Properties  \\\n",
       "141                                                            dataProperty   \n",
       "142                                                            dataProperty   \n",
       "161  1,2,Other,Unknown,Not collected,Not applicable            dataProperty   \n",
       "163                                           1,2,3            dataProperty   \n",
       "169                                                            dataProperty   \n",
       "\n",
       "    Required             Parent DependsOn Component             Source  \\\n",
       "141     True  Biospecimen human                       Sage Bionetworks   \n",
       "142     True  Biospecimen human                       Sage Bionetworks   \n",
       "161     True   Individual Human                       Sage Bionetworks   \n",
       "163     True   Individual Human                       Sage Bionetworks   \n",
       "169     True   Individual Human                      Sage Bionetworks,   \n",
       "\n",
       "                                                 Validation Rules  \n",
       "141  regex search^\\d*?\\.?\\d$|Unknown|Not collected|Not applicable  \n",
       "142  regex search^\\d*?\\.?\\d$|Unknown|Not collected|Not applicable  \n",
       "161                                                                \n",
       "163                                                                \n",
       "169  regex search^\\d*?\\.?\\d$|Unknown|Not collected|Not applicable  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "checks = [c[\"attribute\"] for c in hard_coded_valid_values]\n",
    "\n",
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm2.query(f\"Attribute in @checks\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Schematic config:  C:/Users/nlee/Documents/Projects/schematic/schematic/config.yml \n",
      " CSV model:  C:/Users/nlee/Documents/Projects/ELITE-DCC/ELITE-data-models/models/EL_data_model_v3.1.0.csv \n",
      " JSON LD Model:  C:/Users/nlee/Documents/Projects/ELITE-DCC/ELITE-data-models/models/EL_data_model_v3.jsonld\n"
     ]
    }
   ],
   "source": [
    "with open(\"../configs/notebook_config.yaml\", \"r\") as f:\n",
    "    config = yaml.safe_load(f)\n",
    "\n",
    "# paths to import files\n",
    "root_path = config[\"paths\"][\"root\"]\n",
    "schematic_config = config[\"paths\"][\"schematic\"]\n",
    "csv_model = config[\"file_names\"][\"csv_model\"]\n",
    "json_model = config[\"file_names\"][\"json_model\"]\n",
    "\n",
    "print(\n",
    "    \"Schematic config: \",\n",
    "    schematic_config,\n",
    "    \"\\n\",\n",
    "    \"CSV model: \",\n",
    "    csv_model,\n",
    "    \"\\n\",\n",
    "    \"JSON LD Model: \",\n",
    "    json_model,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Attribute',\n",
       " 'Description',\n",
       " 'Valid Values',\n",
       " 'DependsOn',\n",
       " 'Properties',\n",
       " 'Required',\n",
       " 'Parent',\n",
       " 'DependsOn Component',\n",
       " 'Source',\n",
       " 'Validation Rules']"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm2.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'dm2' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp\\ipykernel_3212\\809299604.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m# write out data model into csv\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mdm2\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcsv_model\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'dm2' is not defined"
     ]
    }
   ],
   "source": [
    "# write out data model into csv\n",
    "dm2.to_csv(csv_model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Convert CSV to JSON LD\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "schematic schema convert C:/Users/nlee/Documents/Projects/ELITE-DCC/ELITE-data-models/models/EL_data_model_v3.1.0.csv --output_jsonld C:/Users/nlee/Documents/Projects/ELITE-DCC/ELITE-data-models/models/EL_data_model_v3.1.0.jsonld\n",
      "^C\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Data Model was created and saved to 'C:/Users/nlee/Documents/Projects/ELITE-DCC/ELITE-data-models/models/EL_data_model_v3.1.0.jsonld' location.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Starting schematic...\n",
      "Done adding requirements and value ranges to attributes\n"
     ]
    }
   ],
   "source": [
    "print(f'schematic schema convert {csv_model} --output_jsonld {json_model}')\n",
    "\n",
    "!schematic schema convert {csv_model} --output_jsonld {json_model}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get manifest names to generate manifests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "with open(json_model, \"r\") as jf:\n",
    "    jo = json.load(jf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Biospecimenhuman': 'Biospecimen human',\n",
       " 'BiospecimennonHuman': 'Biospecimen nonHuman',\n",
       " 'IndividualHuman': 'Individual Human',\n",
       " 'IndividualnonHuman': 'Individual nonHuman',\n",
       " 'MetabolomicsHuman': 'Metabolomics Human',\n",
       " 'Microbiome': 'Microbiome',\n",
       " 'RNAseq': 'RNAseq',\n",
       " 'WholeGenomeSequencing': 'Whole Genome Sequencing',\n",
       " 'BsSeq(bisulfite-seqWGBSmethylseqmethylomics)': 'bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       " 'Genotyping': 'genotyping',\n",
       " 'Proteomics': 'proteomics',\n",
       " 'ScRNAseq': 'scRNAseq'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Manifest names in data model\n",
    "manifest_names_extracted = []\n",
    "\n",
    "for i in jo[\"@graph\"]:\n",
    "    if i[\"@id\"] == \"bts:dataType\":\n",
    "        manifest_names_extracted.append(\n",
    "            i[\"schema:domainIncludes\"][\"@id\"].replace(\"bts:\", \"\")\n",
    "        )\n",
    "\n",
    "# display names extracted\n",
    "manifest_display_names_extracted = []\n",
    "\n",
    "for i in jo[\"@graph\"]:\n",
    "    if i[\"@id\"].replace(\"bts:\", \"\") in (manifest_names_extracted):\n",
    "        manifest_display_names_extracted.append(i[\"sms:displayName\"])\n",
    "manifest_display_names_extracted\n",
    "\n",
    "# Create dictionary for lookup later\n",
    "manifest_name_relationships = dict(\n",
    "    zip(manifest_names_extracted, manifest_display_names_extracted)\n",
    ")\n",
    "\n",
    "manifest_name_relationships"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create dca template configuration\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"manifest_schemas\": [\n",
      "    {\n",
      "      \"display_name\": \"Biospecimen human\",\n",
      "      \"schema_name\": \"Biospecimenhuman\",\n",
      "      \"type\": \"records\"\n",
      "    },\n",
      "    {\n",
      "      \"display_name\": \"Biospecimen nonHuman\",\n",
      "      \"schema_name\": \"BiospecimennonHuman\",\n",
      "      \"type\": \"records\"\n",
      "    },\n",
      "    {\n",
      "      \"display_name\": \"Individual Human\",\n",
      "      \"schema_name\": \"IndividualHuman\",\n",
      "      \"type\": \"records\"\n",
      "    },\n",
      "    {\n",
      "      \"display_name\": \"Individual nonHuman\",\n",
      "      \"schema_name\": \"IndividualnonHuman\",\n",
      "      \"type\": \"records\"\n",
      "    },\n",
      "    {\n",
      "      \"display_name\": \"Metabolomics Human\",\n",
      "      \"schema_name\": \"MetabolomicsHuman\",\n",
      "      \"type\": \"file\"\n",
      "    },\n",
      "    {\n",
      "      \"display_name\": \"Microbiome\",\n",
      "      \"schema_name\": \"Microbiome\",\n",
      "      \"type\": \"file\"\n",
      "    },\n",
      "    {\n",
      "      \"display_name\": \"RNAseq\",\n",
      "      \"schema_name\": \"RNAseq\",\n",
      "      \"type\": \"file\"\n",
      "    },\n",
      "    {\n",
      "      \"display_name\": \"Whole Genome Sequencing\",\n",
      "      \"schema_name\": \"WholeGenomeSequencing\",\n",
      "      \"type\": \"file\"\n",
      "    },\n",
      "    {\n",
      "      \"display_name\": \"bsSeq (bisulfite-seq WGBS methylseq methylomics)\",\n",
      "      \"schema_name\": \"BsSeq(bisulfite-seqWGBSmethylseqmethylomics)\",\n",
      "      \"type\": \"file\"\n",
      "    },\n",
      "    {\n",
      "      \"display_name\": \"genotyping\",\n",
      "      \"schema_name\": \"Genotyping\",\n",
      "      \"type\": \"file\"\n",
      "    },\n",
      "    {\n",
      "      \"display_name\": \"proteomics\",\n",
      "      \"schema_name\": \"Proteomics\",\n",
      "      \"type\": \"file\"\n",
      "    },\n",
      "    {\n",
      "      \"display_name\": \"scRNAseq\",\n",
      "      \"schema_name\": \"ScRNAseq\",\n",
      "      \"type\": \"file\"\n",
      "    }\n",
      "  ],\n",
      "  \"service_version\": \"v23.1.1\",\n",
      "  \"schema_version\": \"v1.2\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "def manifest_template(k, v, t=\"file\"):\n",
    "    manifest_template = {\"display_name\": v, \"schema_name\": k, \"type\": t}\n",
    "    return manifest_template\n",
    "\n",
    "\n",
    "dca_template = {\n",
    "    \"manifest_schemas\": [],\n",
    "    \"service_version\": \"v23.1.1\",\n",
    "    \"schema_version\": \"v1.2\",\n",
    "}\n",
    "\n",
    "records = [\n",
    "    \"IndividualHuman\",\n",
    "    \"IndividualnonHuman\",\n",
    "    \"Biospecimenhuman\",\n",
    "    \"BiospecimennonHuman\",\n",
    "]\n",
    "\n",
    "\n",
    "for k, v in manifest_name_relationships.items():\n",
    "    if k in records:\n",
    "        t = \"records\"\n",
    "    else:\n",
    "        t = \"file\"\n",
    "    dca_template[\"manifest_schemas\"].append(manifest_template(k, v, t))\n",
    "\n",
    "json_formatted_str = json.dumps(dca_template, indent=2)\n",
    "print(json_formatted_str)\n",
    "\n",
    "with open(\n",
    "    \"C:/Users/nlee/Documents/Projects/ELITE-DCC/ELITE-data-models/manifests/dca_template.json\",\n",
    "    \"w\",\n",
    ") as f:\n",
    "    f.write(json_formatted_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# old_schema = {\n",
    "#     \"manifest_schemas\": [\n",
    "#         {\n",
    "#             \"display_name\": \"Biospecimen human\",\n",
    "#             \"schema_name\": \"Biospecimenhuman\",\n",
    "#             \"type\": \"record\"\n",
    "#         },\n",
    "#         {\n",
    "#             \"display_name\": \"Biospecimen nonHuman\",\n",
    "#             \"schema_name\": \"BiospecimennonHuman\",\n",
    "#             \"type\": \"record\"\n",
    "#         },\n",
    "#         {\n",
    "#             \"display_name\": \"Individual Human\",\n",
    "#             \"schema_name\": \"IndividualHuman\",\n",
    "#             \"type\": \"record\"\n",
    "#         },\n",
    "#         {\n",
    "#             \"display_name\": \"Individual nonHuman\",\n",
    "#             \"schema_name\": \"IndividualnonHuman\",\n",
    "#             \"type\": \"record\"\n",
    "#         },\n",
    "#         {\n",
    "#             \"display_name\": \"Metabolomics Human\",\n",
    "#             \"schema_name\": \"MetabolomicsHuman\",\n",
    "#             \"type\": \"file\"\n",
    "#         },\n",
    "#         {\n",
    "#             \"display_name\": \"Microbiome\",\n",
    "#             \"schema_name\": \"Microbiome\",\n",
    "#             \"type\": \"file\"\n",
    "#         },\n",
    "#         {\n",
    "#             \"display_name\": \"RNAseq\",\n",
    "#             \"schema_name\": \"RNAseq\",\n",
    "#             \"type\": \"file\"\n",
    "#         },\n",
    "#         {\n",
    "#             \"display_name\": \"Study\",\n",
    "#             \"schema_name\": \"Study\",\n",
    "#             \"type\": \"file\"\n",
    "#         },\n",
    "#         {\n",
    "#             \"display_name\": \"Whole Genome Sequencing\",\n",
    "#             \"schema_name\": \"WholeGenomeSequencing\",\n",
    "#             \"type\": \"file\"\n",
    "#         },\n",
    "#         {\n",
    "#             \"display_name\": \"'bsSeq (bisulfite-seq WGBS methylseq methylomics)\",\n",
    "#             \"schema_name\": \"'BsSeq(bisulfite-seqWGBSmethylseqmethylomics)\",\n",
    "#             \"type\": \"file\"\n",
    "#         },\n",
    "#         {\n",
    "#             \"display_name\": \"scRNAseq\",\n",
    "#             \"schema_name\": \"ScRNAseq\",\n",
    "#             \"type\": \"file\"\n",
    "#         },\n",
    "#         {\n",
    "#             \"display_name\": \"Genotyping\",\n",
    "#             \"schema_name\": \"Genotyping\",\n",
    "#             \"type\": \"file\"\n",
    "#         },\n",
    "#         {\n",
    "#             \"display_name\": \"Proteomics\",\n",
    "#             \"schema_name\": \"Proteomics\",\n",
    "#             \"type\": \"file\"\n",
    "#         }\n",
    "#     ],\n",
    "#     \"service_version\": \"v23.1.1\",\n",
    "#     \"schema_version\": \"v1.2\"\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(dca_template[\"manifest_schemas\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# len(old_schema['manifest_schemas'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate Manifests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path = \"C:/Users/nlee/Documents/Projects/ELITE-DCC/ELITE-data-models/manifests/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- valueErrors =\n",
    "  - BsSeq(bisulfite-seqWGBSmethylseqmethylomics): [valueReported]\n",
    "  - \"RNAseq\": [rnaBatchID, specifyLibraryPrep]\n",
    "  - 'MetabolomicsHuman': ['specifyLibraryVersion']\n",
    "  - Proteomics : ['specifyTranscriptType']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Remove `(not listed for purposes of this RFC)`\n",
    "- Change Assay to be blank or list of values\n",
    "- Weight needs to be blank\n",
    "- Uknown spelling error\n",
    "- 'False / 'True -> remove '\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp\\ipykernel_3212\\2652145402.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[1;32mwith\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0moption_context\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"display.max_colwidth\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      2\u001b[0m     \u001b[0mdisplay\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdm2\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mquery\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'Attribute == \"ethnicity\"'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm2.query('Attribute == \"ethnicity\"'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "schematic manifest --config C:/Users/nlee/Documents/Projects/schematic/schematic/config.yml get -dt IndividualHuman --output_csv C:/Users/nlee/Documents/Projects/ELITE-DCC/ELITE-data-models/manifests/IndividualHuman.csv --title EL_Manifest_IndividualHuman --sheet_url 2>&1 >> manifest_generation_results.txt\n",
      "Starting schematic...\n",
      "The (model > input > location) argument with value 'C:/Users/nlee/Documents/Projects/ELITE-DCC/ELITE-data-models/models/EL_data_model_v3.1.0.jsonld' is being read from the config file.\n",
      "The '--jsonld' argument is being taken from configuration file (model > input > location), i.e., 'C:/Users/nlee/Documents/Projects/ELITE-DCC/ELITE-data-models/models/EL_data_model_v3.1.0.jsonld'.\n",
      "JSON schema successfully generated from schema.org schema!\n",
      "JSON schema file log stored as C:/Users/nlee/Documents/Projects/ELITE-DCC/ELITE-data-models/models/EL_data_model_v3.1.0.IndividualHuman.schema.json\n",
      "Permission Id: anyoneWithLink\n",
      "Find the manifest template using this Google Sheet URL:\n",
      "Find the manifest template using this CSV file path: C:/Users/nlee/Documents/Projects/ELITE-DCC/ELITE-data-models/manifests/IndividualHuman.xlsx\n"
     ]
    }
   ],
   "source": [
    "manifest = 'IndividualHuman'\n",
    "\n",
    "print(f\"schematic manifest --config {schematic_config} get -dt {manifest} --output_csv {os.path.join(output_path, manifest + '.csv')} --title {'EL_Manifest_' + manifest} --sheet_url 2>&1 >> manifest_generation_results.txt\")\n",
    "\n",
    "!echo '{manifest}:' >> manifest_generation_results.txt\n",
    "!schematic manifest --config {schematic_config} get -dt {manifest} --output_csv {os.path.join(output_path, manifest + '.xlsx')} --title {'EL_Manifest_' + manifest} --sheet_url 2>&1 >> manifest_generation_results.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!echo ----- newly_generated_manifests ------- 2>&1 >> manifest_generation_links.txt\n",
    "for manifest in manifest_names_extracted:\n",
    "    print(manifest)\n",
    "    !echo {manifest}: >> manifest_generation_results.txt\n",
    "    !schematic manifest --config {schematic_config} get -dt {manifest} --output_xlsx {os.path.join(output_path, manifest + '.xlsx')} --title {'EL_Manifest_' + manifest} --sheet_url 2>&1 >> manifest_generation_results.txt\n",
    "    !echo ----------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add validation rules\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mixed_validation_rule(df, attribute):\n",
    "    ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validation of manifests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import string\n",
    "from pathlib import Path\n",
    "import lorem"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hard coded variables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of rows to fill in\n",
    "nrows = 10\n",
    "random.seed = 27\n",
    "\n",
    "# last minute chaos\n",
    "chaos = False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Functions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def valid_values_to_list(df, attribute):\n",
    "    valid_values = (\n",
    "        df.query(f'Attribute == \"{attribute}\"')[\"Valid Values\"].str.split(\",\").values[0]\n",
    "    )\n",
    "\n",
    "    return valid_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_random_value(list_of_vv):\n",
    "    return random.choice(list_of_vv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rand_integer(min=0, max=100):\n",
    "    return random.randint(min, max)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rand_float(min=0, max=100):\n",
    "    return round(random.uniform(0.0, 100.0), 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_random_string():\n",
    "    t = lorem.sentence().split(\" \")[0]\n",
    "    return t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def introduce_random_NAs(df, N=5):\n",
    "    \"\"\"Another test to see if columns can handle empty values or if they will flag the empty value\"\"\"\n",
    "\n",
    "    rows, cols = df.shape\n",
    "\n",
    "    row_index = [get_rand_integer(max=rows - 1) for _ in range(N)]\n",
    "\n",
    "    col_index = [get_rand_integer(max=cols - 1) for _ in range(N)]\n",
    "\n",
    "    indexes = list(zip(row_index, col_index))\n",
    "\n",
    "    # Print indexes to check where values got replaced\n",
    "    print(indexes)\n",
    "    # for i in indexes\n",
    "    df.iloc[row_index, col_index] = np.NaN\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find attribute column, fill in with value\n",
    "def fill_in_attribute(df, index, attribute, value):\n",
    "    df.loc[index, attribute] = value\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gen_mixed_string_with_length(N=12):\n",
    "    # initializing size of string\n",
    "\n",
    "    # using random.choices()\n",
    "    # generating random strings\n",
    "    res = \"\".join(random.choices(string.ascii_uppercase + string.digits, k=N))\n",
    "\n",
    "    # print result\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def random_change():\n",
    "    # list of functions to choose from\n",
    "    choices = [\n",
    "        introduce_random_NAs,\n",
    "        gen_mixed_string_with_length,\n",
    "        get_rand_integer,\n",
    "        get_random_string,\n",
    "        get_rand_float,\n",
    "    ]\n",
    "\n",
    "    choice = random.choice(choices)\n",
    "\n",
    "    print(choice.__name__)\n",
    "\n",
    "    return choice()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def partition(list_in, n):\n",
    "    random.shuffle(list_in)\n",
    "    return [list_in[i::n] for i in range(n)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create individual and biospecimen ids from random text\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for individaul IDs and specimenIDs\n",
    "individaulIds = [gen_mixed_string_with_length(N=5) for _ in range(1, 100)]\n",
    "specimenIds = [gen_mixed_string_with_length() for _ in range(1, 1000)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "specimenIds_partitioned = partition(specimenIds, 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ind_bio_map = []\n",
    "\n",
    "for i, v in enumerate(individaulIds):\n",
    "    for s in specimenIds_partitioned[i]:\n",
    "        ind_bio_map.append({\"individualID\": v, \"specimenID\": s})\n",
    "\n",
    "ind_bio_map = pd.DataFrame(ind_bio_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manifests\n",
    "# Get all the RFC file paths\n",
    "manifest_paths = glob(\n",
    "    r\"C:\\Users\\nlee\\Documents\\Projects\\ELITE-DCC\\ELITE-data-models\\manifests\\*.csv\"\n",
    ")\n",
    "manifest_paths"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Tested:\n",
    "\n",
    "- scRNAseq : Passed (Need to add validation for biospecimenId)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data model\n",
    "data_model_path = r\"C:\\Users\\nlee\\Documents\\Projects\\ELITE-DCC\\ELITE-data-models\\models\\EL_data_model_v3.csv\"\n",
    "\n",
    "dm = pd.read_csv(data_model_path).iloc[:, 1:].fillna(\"\")\n",
    "\n",
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load manifest\n",
    "manifest_path = \"C:\\\\Users\\\\nlee\\\\Documents\\\\Projects\\\\ELITE-DCC\\\\ELITE-data-models\\\\manifests\\\\WholeGenomeSequencing.csv\"\n",
    "\n",
    "df = pd.read_csv(manifest_path)\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "manifest_name = Path(manifest_path).stem\n",
    "\n",
    "parent_name = manifest_name_relationships[manifest_name]\n",
    "\n",
    "df[\"Component\"] = manifest_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Attributes from data model\n",
    "attrs_to_fill = dm[dm[\"Attribute\"] == (parent_name)][\"DependsOn\"].values[0].split(\",\")\n",
    "\n",
    "attrs_to_fill"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm_attrs = (\n",
    "    dm[(dm[\"Attribute\"].isin(attrs_to_fill))]\n",
    "    .drop_duplicates(subset=[\"Attribute\"])\n",
    "    .copy()\n",
    ")\n",
    "\n",
    "dm_attrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find values in dm_attrrs with valid values filled in\n",
    "vv_attrs = dm_attrs[dm_attrs[\"Valid Values\"] != \"\"][\"Attribute\"].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fill in values for the mainfest with the data model\n",
    "for attribute in vv_attrs:\n",
    "    for i in range(nrows):\n",
    "        index = i\n",
    "        temp_vv = valid_values_to_list(dm_attrs, attribute)\n",
    "        new_val = get_random_value(temp_vv)\n",
    "        df = fill_in_attribute(df, index, attribute, new_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update df with sample of individual and biospecimen sample ids\n",
    "ind_sample = ind_bio_map.sample(nrows, replace=True)[\n",
    "    [\"individualID\", \"specimenID\"]\n",
    "].reset_index(drop=True)\n",
    "\n",
    "df.update(ind_sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "swap_validation_coder = {v: k for k, v in validation_coder.items()}\n",
    "\n",
    "# column coding for values\n",
    "dm_attrs[\"swapper\"] = (\n",
    "    dm_attrs[\"Validation Rules\"].fillna(\"number\").replace(swap_validation_coder)\n",
    ")\n",
    "\n",
    "dm_attrs[\"swapper\"] = dm_attrs[\"swapper\"].apply(\n",
    "    lambda x: re.sub(\"^\\d*?\\.?\\d$\", \"number\", x)\n",
    ")\n",
    "\n",
    "# need a better regex later\n",
    "dm_attrs[\"swapper\"] = dm_attrs[\"swapper\"].apply(\n",
    "    lambda x: re.sub(\n",
    "        \"number numbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumbernumber\",\n",
    "        \"number\",\n",
    "        x,\n",
    "    )\n",
    ")\n",
    "\n",
    "print(dm_attrs[\"swapper\"].unique())\n",
    "\n",
    "dm_attrs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "free_form_attrs = dm_attrs[\n",
    "    (dm_attrs[\"Valid Values\"] == \"\")\n",
    "    & (~dm_attrs[\"Attribute\"].isin([\"individualID\", \"specimenID\"]))\n",
    "]\n",
    "\n",
    "free_form_attrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_values(value):\n",
    "    if value == 'string':\n",
    "        # generate string\n",
    "        return get_random_string()\n",
    "    elif value == 'number':\n",
    "        # generate random number\n",
    "        return get_rand_integer()\n",
    "    elif value == 'mixed'\n",
    "        # split string into regex expressions i.e. numbers and valid values\n",
    "        choices = ['string', 'number']\n",
    "        # chose random to fill in cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = \"([0-9]+\\\\.[0-9]*.)|([0-9]+)\"\n",
    "test2 = \"^\\\\d*?\\\\.?\\\\d$|Other|Unknown|Not collected|Not applicable\"\n",
    "\n",
    "# def match_pattern(rule):\n",
    "\n",
    "pattern_checker = {\n",
    "    \"strip_patterns\": \"regex|search|match\",\n",
    "    \"number\": \"(?![a-zA-Z]+)(\\[0-9\\])\",\n",
    "    \"string\": \"([a-zA-Z]+)(?!\\[0-9\\])\",\n",
    "    \"mixed\": \"(?:[a-zA-Z]+\\d+)\",\n",
    "}\n",
    "\n",
    "result = False\n",
    "\n",
    "while result == False:\n",
    "    for k, v in pattern_checker.items():\n",
    "        result = bool(re.search(pattern=v, string=test))\n",
    "        print(pattern)\n",
    "        print(result)\n",
    "\n",
    "\n",
    "# pattern = '(?![a-zA-Z]+)(\\[0-9\\])'\n",
    "# results = bool(re.search(pattern = pattern, string = test))\n",
    "# results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for attribute in free_form_attrs[\"Attribute\"].tolist():\n",
    "    for i in range(nrows):\n",
    "        index = i\n",
    "        test = free_form_attrs.query(f'Attribute == \"{attribute}\"')\n",
    "        new_val = test[\"swapper\"].apply(generate_values).values[0]\n",
    "        df = fill_in_attribute(df, index, attribute, new_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out test manifest to file for testing in DCA\n",
    "df[\"Component\"] = manifest_name\n",
    "\n",
    "if chaos:\n",
    "    df = introduce_random_NAs(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write out manifest\n",
    "csv_path = os.path.join(\n",
    "    r\"C:\\Users\\nlee\\Documents\\Projects\\ELITE-DCC\\ELITE-data-models\\test_manifests\",\n",
    "    manifest_name + \"_test.csv\",\n",
    ")\n",
    "\n",
    "df.to_csv(csv_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run Validation Test\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# manifest_path = \"C:/Users/nlee/Documents/Projects/schematic/schematic/tests/data/mock_manifests/example_biospecimen_test.csv\"\n",
    "print(\n",
    "    f\"schematic model --config {schematic_config} validate --manifest_path {csv_path} --data_type {manifest_name}\")\n",
    "!schematic model --config {schematic_config} validate --manifest_path {csv_path} --data_type {manifest_name}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Submit Manifest\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(csv_path)\n",
    "print(schematic_config)\n",
    "print(manifest_name)\n",
    "dataset_id = \"syn51753850\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- d option is the project\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_id = 'syn51753844'\n",
    "\n",
    "!schematic model --config C:/Users/nlee/Documents/Projects/schematic/schematic/config.yml submit -mp C:/Users/nlee/Documents/Projects/ELITE-DCC/ELITE-data-models/test_manifests/WholeGenomeSequencing_test.csv -d syn51753850 -vc WholeGenomeSequencing -mrt both"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "search_df(dm2, \"valueReported\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Store Manifest In Dataset For Testing\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import synapseclient\n",
    "from synapseclient import File\n",
    "\n",
    "manifest_folder_id = 'syn51728840'\n",
    "\n",
    "%run C:\\Users\\nlee\\Documents\\Projects\\utils\\syanpse_login.py\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "csv_entity = File(\n",
    "    csv_path,\n",
    "    description=f\"Test manifest for {manifest_name}\",\n",
    "    parent=data_folder,\n",
    "    annotations={\"resourceType\": \"manifest\", \"manifestType\": {manifest_name}},\n",
    ")\n",
    "\n",
    "csv_entity = syn.store(csv_entity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = syn.getColumns(\"syn51748558\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in s:\n",
    "    print(i)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
